// java_/ConstructorInvocationBase.ts
import {Set} from "ecore/Set";
import {AbstractCollection} from "ecore/AbstractCollection";
import {ENotificationImpl} from "ecore/ENotificationImpl";
import {BasicEObjectImpl} from "ecore/BasicEObjectImpl";
import {Expression} from "java_/Expression";
import {EClass} from "ecore/EClass";
import {AbstractMethodInvocationImpl} from "java_/AbstractMethodInvocationImpl";
import {ConstructorInvocation} from "java_/ConstructorInvocation";
import {InternalEObject} from "ecore/InternalEObject";
import {Java_PackageLiterals} from "java_/Java_PackageLiterals";
import {OrderedSet} from "ecore/OrderedSet";
import {EObject} from "ecore/EObject";
import {NotificationImpl} from "ecore/NotificationImpl";
import {AbstractMethodDeclaration} from "java_/AbstractMethodDeclaration";
import {TypeAccess} from "java_/TypeAccess";
import {NotificationChain} from "ecore/NotificationChain";
import {StatementImpl} from "java_/StatementImpl";
		
			export class ConstructorInvocationBase
			extends StatementImpl
			implements ConstructorInvocation
			{
				private _method:AbstractMethodDeclaration = null;
				get method():AbstractMethodDeclaration{
				
					return this._method;
				}
				set method(value:AbstractMethodDeclaration) {
					if (value != this._method) {
						let msgs:NotificationChain = null;
						if (this._method != null){
							msgs = (this._method).eInverseRemove(this, Java_PackageLiterals.ABSTRACTMETHODDECLARATION_USAGES, /*AbstractMethodInvocation*/ null , msgs);
						}
						if (value != null){
							msgs = value.eInverseAdd(this, Java_PackageLiterals.ABSTRACTMETHODDECLARATION_USAGES, /*AbstractMethodInvocation*/ null, msgs);
						}
						msgs = this.basicSetMethod(value, msgs);
						if (msgs != null) {
							msgs.dispatch();
						}
					}
					else if (this.eNotificationRequired()){
						this.eNotify(new ENotificationImpl(this, NotificationImpl.SET,Java_PackageLiterals.ABSTRACTMETHODINVOCATION_METHOD , value, value));
					}
				}
				private _arguments:OrderedSet<Expression> = null;
				get arguments():OrderedSet<Expression>{
					if(this._arguments===null){
						this._arguments = new OrderedSet<Expression>(this, Java_PackageLiterals.ABSTRACTMETHODINVOCATION_ARGUMENTS, BasicEObjectImpl.EOPPOSITE_FEATURE_BASE - Java_PackageLiterals.ABSTRACTMETHODINVOCATION_ARGUMENTS);
							
					}
					return this._arguments;
					
				}
				
				
				private _typeArguments:OrderedSet<TypeAccess> = null;
				get typeArguments():OrderedSet<TypeAccess>{
					if(this._typeArguments===null){
						this._typeArguments = new OrderedSet<TypeAccess>(this, Java_PackageLiterals.ABSTRACTMETHODINVOCATION_TYPEARGUMENTS, BasicEObjectImpl.EOPPOSITE_FEATURE_BASE - Java_PackageLiterals.ABSTRACTMETHODINVOCATION_TYPEARGUMENTS);
							
					}
					return this._typeArguments;
					
				}
				
				

			
				public static eStaticClass:EClass;
				
				protected eStaticClass():EClass{
					
					return ConstructorInvocationBase.eStaticClass;
				}
			
				public eInverseAdd(otherEnd:InternalEObject, featureID:number, msgs:NotificationChain): NotificationChain{
					switch (featureID) {
						case Java_PackageLiterals.CONSTRUCTORINVOCATION_METHOD:
							if (this._method != null){
								msgs = this._method.eInverseRemove(this, Java_PackageLiterals.ABSTRACTMETHODINVOCATION_METHOD, /*AbstractMethodDeclaration*/ null, msgs);
							}
							return this.basicSetMethod(otherEnd as AbstractMethodDeclaration, msgs);
					}
					//return this.eInverseAddFromStatement(otherEnd, featureID, msgs);
					return super.eInverseAdd(otherEnd, featureID, msgs);
				}
				//public eInverseAddFromConstructorInvocation = this.eInverseAdd;
				
				public eInverseRemove(otherEnd:InternalEObject, featureID:number, msgs:NotificationChain):NotificationChain{
					switch (featureID) {
						case Java_PackageLiterals.CONSTRUCTORINVOCATION_METHOD:
							return this.basicSetMethod(null, msgs);
					}
					//return this.eInverseRemoveFromStatement(otherEnd, featureID, msgs);
					return super.eInverseRemove(otherEnd, featureID, msgs);
				}
				
				//public eInverseRemoveFromConstructorInvocation = this.eInverseRemove;
			
				public basicSetMethod(newobj:AbstractMethodDeclaration, msgs:NotificationChain):NotificationChain {
					let oldobj = this._method;
					this._method = newobj;
					if (this.eNotificationRequired()) {
						let notification = new ENotificationImpl(this, NotificationImpl.SET, Java_PackageLiterals.ABSTRACTMETHODINVOCATION_METHOD, oldobj, newobj);
						if (msgs == null){
							msgs = notification;
						}
						else{
							msgs.add(notification);
						}
					}
					return msgs;
				}
				
			
				public eGet_number_boolean_boolean(featureID:number, resolve:boolean, coreType:boolean):any{
					switch (featureID) {
						case Java_PackageLiterals.CONSTRUCTORINVOCATION_COMMENTS:
							return this.comments;
						case Java_PackageLiterals.CONSTRUCTORINVOCATION_ORIGINALCOMPILATIONUNIT:
							return this.originalCompilationUnit;
						case Java_PackageLiterals.CONSTRUCTORINVOCATION_ORIGINALCLASSFILE:
							return this.originalClassFile;
						case Java_PackageLiterals.CONSTRUCTORINVOCATION_METHOD:
							return this.method;
						case Java_PackageLiterals.CONSTRUCTORINVOCATION_ARGUMENTS:
							return this.arguments;
						case Java_PackageLiterals.CONSTRUCTORINVOCATION_TYPEARGUMENTS:
							return this.typeArguments;
					}
					//return this.eGetFromStatement(featureID, resolve, coreType);
					return super.eGet(featureID, resolve, coreType);
				}
				
				public eSet_number_any(featureID:number, newValue:any):void {
					switch (featureID) {
						case Java_PackageLiterals.CONSTRUCTORINVOCATION_METHOD:
							this.method = <AbstractMethodDeclaration> newValue;
							return;
						case Java_PackageLiterals.CONSTRUCTORINVOCATION_ARGUMENTS:
							this.arguments.clear();
							this.arguments.addAll(newValue);
							return;
						case Java_PackageLiterals.CONSTRUCTORINVOCATION_TYPEARGUMENTS:
							this.typeArguments.clear();
							this.typeArguments.addAll(newValue);
							return;
					}
					super.eSet_number_any(featureID, newValue);
				}

				public eBaseStructuralFeatureID(derivedFeatureID:number, baseClass:Function):number {
					if (baseClass === AbstractMethodInvocationImpl) {
						switch (derivedFeatureID) {
							case Java_PackageLiterals.CONSTRUCTORINVOCATION_METHOD: return Java_PackageLiterals.ABSTRACTMETHODINVOCATION_METHOD;
							case Java_PackageLiterals.CONSTRUCTORINVOCATION_ARGUMENTS: return Java_PackageLiterals.ABSTRACTMETHODINVOCATION_ARGUMENTS;
							case Java_PackageLiterals.CONSTRUCTORINVOCATION_TYPEARGUMENTS: return Java_PackageLiterals.ABSTRACTMETHODINVOCATION_TYPEARGUMENTS;
							default: return -1;
						}
					}
					return super.eBaseStructuralFeatureID(derivedFeatureID, baseClass);
				}
				
							
				public eDerivedStructuralFeatureID_number_Function(baseFeatureID:number, baseClass:Function):number {
					if (baseClass === AbstractMethodInvocationImpl) {
						switch (baseFeatureID) {
							case Java_PackageLiterals.ABSTRACTMETHODINVOCATION_METHOD: return Java_PackageLiterals.CONSTRUCTORINVOCATION_METHOD;
							case Java_PackageLiterals.ABSTRACTMETHODINVOCATION_ARGUMENTS: return Java_PackageLiterals.CONSTRUCTORINVOCATION_ARGUMENTS;
							case Java_PackageLiterals.ABSTRACTMETHODINVOCATION_TYPEARGUMENTS: return Java_PackageLiterals.CONSTRUCTORINVOCATION_TYPEARGUMENTS;
							default: return -1;
						}
					}
					return super.eDerivedStructuralFeatureID_number_Function(baseFeatureID, baseClass);
				}
				
			}
