// java_/Java_Switch.ts
import {ForStatement} from "java_/ForStatement";
import {AnnotationMemberValuePair} from "java_/AnnotationMemberValuePair";
import {InstanceofExpression} from "java_/InstanceofExpression";
import {MemberRef} from "java_/MemberRef";
import {CompilationUnit} from "java_/CompilationUnit";
import {PrimitiveTypeShort} from "java_/PrimitiveTypeShort";
import {WildCardType} from "java_/WildCardType";
import {MethodInvocation} from "java_/MethodInvocation";
import {EPackage} from "ecore/EPackage";
import {VariableDeclarationFragment} from "java_/VariableDeclarationFragment";
import {ConstructorDeclaration} from "java_/ConstructorDeclaration";
import {SuperConstructorInvocation} from "java_/SuperConstructorInvocation";
import {ContinueStatement} from "java_/ContinueStatement";
import {AbstractMethodInvocation} from "java_/AbstractMethodInvocation";
import {AbstractTypeDeclaration} from "java_/AbstractTypeDeclaration";
import {TextElement} from "java_/TextElement";
import {SwitchCase} from "java_/SwitchCase";
import {Statement} from "java_/Statement";
import {TypeDeclaration} from "java_/TypeDeclaration";
import {ClassFile} from "java_/ClassFile";
import {TypeParameter} from "java_/TypeParameter";
import {PrimitiveTypeVoid} from "java_/PrimitiveTypeVoid";
import {ThrowStatement} from "java_/ThrowStatement";
import {ThisExpression} from "java_/ThisExpression";
import {PostfixExpression} from "java_/PostfixExpression";
import {AbstractVariablesContainer} from "java_/AbstractVariablesContainer";
import {EnumConstantDeclaration} from "java_/EnumConstantDeclaration";
import {Archive} from "java_/Archive";
import {LineComment} from "java_/LineComment";
import {Package} from "java_/Package";
import {UnresolvedClassDeclaration} from "java_/UnresolvedClassDeclaration";
import {Java_PackageImpl} from "java_/Java_PackageImpl";
import {PrimitiveTypeLong} from "java_/PrimitiveTypeLong";
import {SingleVariableDeclaration} from "java_/SingleVariableDeclaration";
import {Modifier} from "java_/Modifier";
import {ArrayLengthAccess} from "java_/ArrayLengthAccess";
import {MethodDeclaration} from "java_/MethodDeclaration";
import {UnresolvedAnnotationTypeMemberDeclaration} from "java_/UnresolvedAnnotationTypeMemberDeclaration";
import {VariableDeclarationExpression} from "java_/VariableDeclarationExpression";
import {ManifestEntry} from "java_/ManifestEntry";
import {NullLiteral} from "java_/NullLiteral";
import {ParameterizedType} from "java_/ParameterizedType";
import {LabeledStatement} from "java_/LabeledStatement";
import {PrimitiveTypeDouble} from "java_/PrimitiveTypeDouble";
import {EObject} from "ecore/EObject";
import {Annotation} from "java_/Annotation";
import {InterfaceDeclaration} from "java_/InterfaceDeclaration";
import {ReturnStatement} from "java_/ReturnStatement";
import {CharacterLiteral} from "java_/CharacterLiteral";
import {ClassInstanceCreation} from "java_/ClassInstanceCreation";
import {EnhancedForStatement} from "java_/EnhancedForStatement";
import {UnresolvedLabeledStatement} from "java_/UnresolvedLabeledStatement";
import {Switch} from "ecore/Switch";
import {FieldAccess} from "java_/FieldAccess";
import {NamedElement} from "java_/NamedElement";
import {PrimitiveTypeChar} from "java_/PrimitiveTypeChar";
import {BodyDeclaration} from "java_/BodyDeclaration";
import {Javadoc} from "java_/Javadoc";
import {UnresolvedEnumDeclaration} from "java_/UnresolvedEnumDeclaration";
import {ArrayCreation} from "java_/ArrayCreation";
import {TypeDeclarationStatement} from "java_/TypeDeclarationStatement";
import {ArrayType} from "java_/ArrayType";
import {UnresolvedInterfaceDeclaration} from "java_/UnresolvedInterfaceDeclaration";
import {NumberLiteral} from "java_/NumberLiteral";
import {AbstractMethodDeclaration} from "java_/AbstractMethodDeclaration";
import {ImportDeclaration} from "java_/ImportDeclaration";
import {PrimitiveTypeFloat} from "java_/PrimitiveTypeFloat";
import {ConditionalExpression} from "java_/ConditionalExpression";
import {VariableDeclarationStatement} from "java_/VariableDeclarationStatement";
import {Block} from "java_/Block";
import {Model} from "java_/Model";
import {BooleanLiteral} from "java_/BooleanLiteral";
import {AbstractTypeQualifiedExpression} from "java_/AbstractTypeQualifiedExpression";
import {AnonymousClassDeclaration} from "java_/AnonymousClassDeclaration";
import {IfStatement} from "java_/IfStatement";
import {UnresolvedAnnotationDeclaration} from "java_/UnresolvedAnnotationDeclaration";
import {Java_Package} from "java_/Java_Package";
import {PrefixExpression} from "java_/PrefixExpression";
import {ClassDeclaration} from "java_/ClassDeclaration";
import {Type} from "java_/Type";
import {ASTNode} from "java_/ASTNode";
import {BreakStatement} from "java_/BreakStatement";
import {UnresolvedItem} from "java_/UnresolvedItem";
import {BlockComment} from "java_/BlockComment";
import {ArrayAccess} from "java_/ArrayAccess";
import {TypeAccess} from "java_/TypeAccess";
import {WhileStatement} from "java_/WhileStatement";
import {CatchClause} from "java_/CatchClause";
import {UnresolvedSingleVariableDeclaration} from "java_/UnresolvedSingleVariableDeclaration";
import {NamespaceAccess} from "java_/NamespaceAccess";
import {CastExpression} from "java_/CastExpression";
import {Expression} from "java_/Expression";
import {ParenthesizedExpression} from "java_/ParenthesizedExpression";
import {SuperFieldAccess} from "java_/SuperFieldAccess";
import {ConstructorInvocation} from "java_/ConstructorInvocation";
import {SingleVariableAccess} from "java_/SingleVariableAccess";
import {MethodRef} from "java_/MethodRef";
import {AnnotationTypeDeclaration} from "java_/AnnotationTypeDeclaration";
import {DoStatement} from "java_/DoStatement";
import {EnumDeclaration} from "java_/EnumDeclaration";
import {SuperMethodInvocation} from "java_/SuperMethodInvocation";
import {StringLiteral} from "java_/StringLiteral";
import {UnresolvedTypeDeclaration} from "java_/UnresolvedTypeDeclaration";
import {Comment} from "java_/Comment";
import {TypeLiteral} from "java_/TypeLiteral";
import {Initializer} from "java_/Initializer";
import {PrimitiveType} from "java_/PrimitiveType";
import {AssertStatement} from "java_/AssertStatement";
import {FieldDeclaration} from "java_/FieldDeclaration";
import {PackageAccess} from "java_/PackageAccess";
import {VariableDeclaration} from "java_/VariableDeclaration";
import {UnresolvedVariableDeclarationFragment} from "java_/UnresolvedVariableDeclarationFragment";
import {ManifestAttribute} from "java_/ManifestAttribute";
import {UnresolvedMethodDeclaration} from "java_/UnresolvedMethodDeclaration";
import {Manifest} from "java_/Manifest";
import {SwitchStatement} from "java_/SwitchStatement";
import {PrimitiveTypeInt} from "java_/PrimitiveTypeInt";
import {SynchronizedStatement} from "java_/SynchronizedStatement";
import {ArrayInitializer} from "java_/ArrayInitializer";
import {PrimitiveTypeByte} from "java_/PrimitiveTypeByte";
import {UnresolvedItemAccess} from "java_/UnresolvedItemAccess";
import {MethodRefParameter} from "java_/MethodRefParameter";
import {TryStatement} from "java_/TryStatement";
import {Assignment} from "java_/Assignment";
import {InfixExpression} from "java_/InfixExpression";
import {AnnotationTypeMemberDeclaration} from "java_/AnnotationTypeMemberDeclaration";
import {PrimitiveTypeBoolean} from "java_/PrimitiveTypeBoolean";
import {EmptyStatement} from "java_/EmptyStatement";
import {UnresolvedType} from "java_/UnresolvedType";
import {TagElement} from "java_/TagElement";
import {ExpressionStatement} from "java_/ExpressionStatement";
export class Java_Switch<T> extends Switch<T> {
	protected static modelPackage:Java_Package;
	
	constructor(){
		super();
		if (Java_Switch.modelPackage == null) {
			Java_Switch.modelPackage = Java_PackageImpl.eINSTANCE;
		}	
	}
	
	public isSwitchFor(ePackage:EPackage):boolean{
		return ePackage === Java_Switch.modelPackage;
	}
	
	public doSwitch(classifierID:number, theEObject:EObject):T {
		switch (classifierID) {
			case Java_PackageImpl.ABSTRACTMETHODDECLARATION: {
				let obj:AbstractMethodDeclaration = <AbstractMethodDeclaration>theEObject;
				let result:T = this.caseAbstractMethodDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ABSTRACTMETHODINVOCATION: {
				let obj:AbstractMethodInvocation = <AbstractMethodInvocation>theEObject;
				let result:T = this.caseAbstractMethodInvocation(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ABSTRACTTYPEDECLARATION: {
				let obj:AbstractTypeDeclaration = <AbstractTypeDeclaration>theEObject;
				let result:T = this.caseAbstractTypeDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ABSTRACTTYPEQUALIFIEDEXPRESSION: {
				let obj:AbstractTypeQualifiedExpression = <AbstractTypeQualifiedExpression>theEObject;
				let result:T = this.caseAbstractTypeQualifiedExpression(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ABSTRACTVARIABLESCONTAINER: {
				let obj:AbstractVariablesContainer = <AbstractVariablesContainer>theEObject;
				let result:T = this.caseAbstractVariablesContainer(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ANNOTATION: {
				let obj:Annotation = <Annotation>theEObject;
				let result:T = this.caseAnnotation(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ARCHIVE: {
				let obj:Archive = <Archive>theEObject;
				let result:T = this.caseArchive(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ASSERTSTATEMENT: {
				let obj:AssertStatement = <AssertStatement>theEObject;
				let result:T = this.caseAssertStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ASTNODE: {
				let obj:ASTNode = <ASTNode>theEObject;
				let result:T = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ANNOTATIONMEMBERVALUEPAIR: {
				let obj:AnnotationMemberValuePair = <AnnotationMemberValuePair>theEObject;
				let result:T = this.caseAnnotationMemberValuePair(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ANNOTATIONTYPEDECLARATION: {
				let obj:AnnotationTypeDeclaration = <AnnotationTypeDeclaration>theEObject;
				let result:T = this.caseAnnotationTypeDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.caseAbstractTypeDeclaration(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION: {
				let obj:AnnotationTypeMemberDeclaration = <AnnotationTypeMemberDeclaration>theEObject;
				let result:T = this.caseAnnotationTypeMemberDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ANONYMOUSCLASSDECLARATION: {
				let obj:AnonymousClassDeclaration = <AnonymousClassDeclaration>theEObject;
				let result:T = this.caseAnonymousClassDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ARRAYACCESS: {
				let obj:ArrayAccess = <ArrayAccess>theEObject;
				let result:T = this.caseArrayAccess(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ARRAYCREATION: {
				let obj:ArrayCreation = <ArrayCreation>theEObject;
				let result:T = this.caseArrayCreation(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ARRAYINITIALIZER: {
				let obj:ArrayInitializer = <ArrayInitializer>theEObject;
				let result:T = this.caseArrayInitializer(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ARRAYLENGTHACCESS: {
				let obj:ArrayLengthAccess = <ArrayLengthAccess>theEObject;
				let result:T = this.caseArrayLengthAccess(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ARRAYTYPE: {
				let obj:ArrayType = <ArrayType>theEObject;
				let result:T = this.caseArrayType(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ASSIGNMENT: {
				let obj:Assignment = <Assignment>theEObject;
				let result:T = this.caseAssignment(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.BODYDECLARATION: {
				let obj:BodyDeclaration = <BodyDeclaration>theEObject;
				let result:T = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.BOOLEANLITERAL: {
				let obj:BooleanLiteral = <BooleanLiteral>theEObject;
				let result:T = this.caseBooleanLiteral(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.BLOCKCOMMENT: {
				let obj:BlockComment = <BlockComment>theEObject;
				let result:T = this.caseBlockComment(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseComment(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.BLOCK: {
				let obj:Block = <Block>theEObject;
				let result:T = this.caseBlock(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.BREAKSTATEMENT: {
				let obj:BreakStatement = <BreakStatement>theEObject;
				let result:T = this.caseBreakStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.CASTEXPRESSION: {
				let obj:CastExpression = <CastExpression>theEObject;
				let result:T = this.caseCastExpression(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.CATCHCLAUSE: {
				let obj:CatchClause = <CatchClause>theEObject;
				let result:T = this.caseCatchClause(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.CHARACTERLITERAL: {
				let obj:CharacterLiteral = <CharacterLiteral>theEObject;
				let result:T = this.caseCharacterLiteral(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.CLASSFILE: {
				let obj:ClassFile = <ClassFile>theEObject;
				let result:T = this.caseClassFile(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.CLASSINSTANCECREATION: {
				let obj:ClassInstanceCreation = <ClassInstanceCreation>theEObject;
				let result:T = this.caseClassInstanceCreation(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.caseAbstractMethodInvocation(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.CONSTRUCTORDECLARATION: {
				let obj:ConstructorDeclaration = <ConstructorDeclaration>theEObject;
				let result:T = this.caseConstructorDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseAbstractMethodDeclaration(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.CONDITIONALEXPRESSION: {
				let obj:ConditionalExpression = <ConditionalExpression>theEObject;
				let result:T = this.caseConditionalExpression(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.CONSTRUCTORINVOCATION: {
				let obj:ConstructorInvocation = <ConstructorInvocation>theEObject;
				let result:T = this.caseConstructorInvocation(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.caseAbstractMethodInvocation(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.CLASSDECLARATION: {
				let obj:ClassDeclaration = <ClassDeclaration>theEObject;
				let result:T = this.caseClassDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.caseAbstractTypeDeclaration(obj);
				if (result == null) result = this.caseTypeDeclaration(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.COMMENT: {
				let obj:Comment = <Comment>theEObject;
				let result:T = this.caseComment(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.COMPILATIONUNIT: {
				let obj:CompilationUnit = <CompilationUnit>theEObject;
				let result:T = this.caseCompilationUnit(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.CONTINUESTATEMENT: {
				let obj:ContinueStatement = <ContinueStatement>theEObject;
				let result:T = this.caseContinueStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.DOSTATEMENT: {
				let obj:DoStatement = <DoStatement>theEObject;
				let result:T = this.caseDoStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.EMPTYSTATEMENT: {
				let obj:EmptyStatement = <EmptyStatement>theEObject;
				let result:T = this.caseEmptyStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ENHANCEDFORSTATEMENT: {
				let obj:EnhancedForStatement = <EnhancedForStatement>theEObject;
				let result:T = this.caseEnhancedForStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ENUMCONSTANTDECLARATION: {
				let obj:EnumConstantDeclaration = <EnumConstantDeclaration>theEObject;
				let result:T = this.caseEnumConstantDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseVariableDeclaration(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.ENUMDECLARATION: {
				let obj:EnumDeclaration = <EnumDeclaration>theEObject;
				let result:T = this.caseEnumDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.caseAbstractTypeDeclaration(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.EXPRESSION: {
				let obj:Expression = <Expression>theEObject;
				let result:T = this.caseExpression(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.EXPRESSIONSTATEMENT: {
				let obj:ExpressionStatement = <ExpressionStatement>theEObject;
				let result:T = this.caseExpressionStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.FIELDACCESS: {
				let obj:FieldAccess = <FieldAccess>theEObject;
				let result:T = this.caseFieldAccess(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.FIELDDECLARATION: {
				let obj:FieldDeclaration = <FieldDeclaration>theEObject;
				let result:T = this.caseFieldDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseAbstractVariablesContainer(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.FORSTATEMENT: {
				let obj:ForStatement = <ForStatement>theEObject;
				let result:T = this.caseForStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.IFSTATEMENT: {
				let obj:IfStatement = <IfStatement>theEObject;
				let result:T = this.caseIfStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.IMPORTDECLARATION: {
				let obj:ImportDeclaration = <ImportDeclaration>theEObject;
				let result:T = this.caseImportDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.INFIXEXPRESSION: {
				let obj:InfixExpression = <InfixExpression>theEObject;
				let result:T = this.caseInfixExpression(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.INITIALIZER: {
				let obj:Initializer = <Initializer>theEObject;
				let result:T = this.caseInitializer(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.INSTANCEOFEXPRESSION: {
				let obj:InstanceofExpression = <InstanceofExpression>theEObject;
				let result:T = this.caseInstanceofExpression(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.INTERFACEDECLARATION: {
				let obj:InterfaceDeclaration = <InterfaceDeclaration>theEObject;
				let result:T = this.caseInterfaceDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.caseAbstractTypeDeclaration(obj);
				if (result == null) result = this.caseTypeDeclaration(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.JAVADOC: {
				let obj:Javadoc = <Javadoc>theEObject;
				let result:T = this.caseJavadoc(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseComment(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.LABELEDSTATEMENT: {
				let obj:LabeledStatement = <LabeledStatement>theEObject;
				let result:T = this.caseLabeledStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.LINECOMMENT: {
				let obj:LineComment = <LineComment>theEObject;
				let result:T = this.caseLineComment(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseComment(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.MANIFEST: {
				let obj:Manifest = <Manifest>theEObject;
				let result:T = this.caseManifest(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.MANIFESTATTRIBUTE: {
				let obj:ManifestAttribute = <ManifestAttribute>theEObject;
				let result:T = this.caseManifestAttribute(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.MANIFESTENTRY: {
				let obj:ManifestEntry = <ManifestEntry>theEObject;
				let result:T = this.caseManifestEntry(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.MEMBERREF: {
				let obj:MemberRef = <MemberRef>theEObject;
				let result:T = this.caseMemberRef(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.METHODDECLARATION: {
				let obj:MethodDeclaration = <MethodDeclaration>theEObject;
				let result:T = this.caseMethodDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseAbstractMethodDeclaration(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.METHODINVOCATION: {
				let obj:MethodInvocation = <MethodInvocation>theEObject;
				let result:T = this.caseMethodInvocation(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.caseAbstractMethodInvocation(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.METHODREF: {
				let obj:MethodRef = <MethodRef>theEObject;
				let result:T = this.caseMethodRef(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.METHODREFPARAMETER: {
				let obj:MethodRefParameter = <MethodRefParameter>theEObject;
				let result:T = this.caseMethodRefParameter(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.MODEL: {
				let obj:Model = <Model>theEObject;
				let result:T = this.caseModel(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.MODIFIER: {
				let obj:Modifier = <Modifier>theEObject;
				let result:T = this.caseModifier(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.NAMEDELEMENT: {
				let obj:NamedElement = <NamedElement>theEObject;
				let result:T = this.caseNamedElement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.NAMESPACEACCESS: {
				let obj:NamespaceAccess = <NamespaceAccess>theEObject;
				let result:T = this.caseNamespaceAccess(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.NUMBERLITERAL: {
				let obj:NumberLiteral = <NumberLiteral>theEObject;
				let result:T = this.caseNumberLiteral(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.NULLLITERAL: {
				let obj:NullLiteral = <NullLiteral>theEObject;
				let result:T = this.caseNullLiteral(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.PACKAGE: {
				let obj:Package = <Package>theEObject;
				let result:T = this.casePackage(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.PACKAGEACCESS: {
				let obj:PackageAccess = <PackageAccess>theEObject;
				let result:T = this.casePackageAccess(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamespaceAccess(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.PARAMETERIZEDTYPE: {
				let obj:ParameterizedType = <ParameterizedType>theEObject;
				let result:T = this.caseParameterizedType(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.PARENTHESIZEDEXPRESSION: {
				let obj:ParenthesizedExpression = <ParenthesizedExpression>theEObject;
				let result:T = this.caseParenthesizedExpression(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.POSTFIXEXPRESSION: {
				let obj:PostfixExpression = <PostfixExpression>theEObject;
				let result:T = this.casePostfixExpression(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.PREFIXEXPRESSION: {
				let obj:PrefixExpression = <PrefixExpression>theEObject;
				let result:T = this.casePrefixExpression(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.PRIMITIVETYPE: {
				let obj:PrimitiveType = <PrimitiveType>theEObject;
				let result:T = this.casePrimitiveType(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.PRIMITIVETYPEBOOLEAN: {
				let obj:PrimitiveTypeBoolean = <PrimitiveTypeBoolean>theEObject;
				let result:T = this.casePrimitiveTypeBoolean(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.casePrimitiveType(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.PRIMITIVETYPEBYTE: {
				let obj:PrimitiveTypeByte = <PrimitiveTypeByte>theEObject;
				let result:T = this.casePrimitiveTypeByte(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.casePrimitiveType(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.PRIMITIVETYPECHAR: {
				let obj:PrimitiveTypeChar = <PrimitiveTypeChar>theEObject;
				let result:T = this.casePrimitiveTypeChar(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.casePrimitiveType(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.PRIMITIVETYPEDOUBLE: {
				let obj:PrimitiveTypeDouble = <PrimitiveTypeDouble>theEObject;
				let result:T = this.casePrimitiveTypeDouble(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.casePrimitiveType(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.PRIMITIVETYPESHORT: {
				let obj:PrimitiveTypeShort = <PrimitiveTypeShort>theEObject;
				let result:T = this.casePrimitiveTypeShort(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.casePrimitiveType(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.PRIMITIVETYPEFLOAT: {
				let obj:PrimitiveTypeFloat = <PrimitiveTypeFloat>theEObject;
				let result:T = this.casePrimitiveTypeFloat(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.casePrimitiveType(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.PRIMITIVETYPEINT: {
				let obj:PrimitiveTypeInt = <PrimitiveTypeInt>theEObject;
				let result:T = this.casePrimitiveTypeInt(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.casePrimitiveType(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.PRIMITIVETYPELONG: {
				let obj:PrimitiveTypeLong = <PrimitiveTypeLong>theEObject;
				let result:T = this.casePrimitiveTypeLong(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.casePrimitiveType(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.PRIMITIVETYPEVOID: {
				let obj:PrimitiveTypeVoid = <PrimitiveTypeVoid>theEObject;
				let result:T = this.casePrimitiveTypeVoid(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.casePrimitiveType(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.RETURNSTATEMENT: {
				let obj:ReturnStatement = <ReturnStatement>theEObject;
				let result:T = this.caseReturnStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.SINGLEVARIABLEACCESS: {
				let obj:SingleVariableAccess = <SingleVariableAccess>theEObject;
				let result:T = this.caseSingleVariableAccess(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.SINGLEVARIABLEDECLARATION: {
				let obj:SingleVariableDeclaration = <SingleVariableDeclaration>theEObject;
				let result:T = this.caseSingleVariableDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseVariableDeclaration(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.STATEMENT: {
				let obj:Statement = <Statement>theEObject;
				let result:T = this.caseStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.STRINGLITERAL: {
				let obj:StringLiteral = <StringLiteral>theEObject;
				let result:T = this.caseStringLiteral(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.SUPERCONSTRUCTORINVOCATION: {
				let obj:SuperConstructorInvocation = <SuperConstructorInvocation>theEObject;
				let result:T = this.caseSuperConstructorInvocation(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.caseAbstractMethodInvocation(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.SUPERFIELDACCESS: {
				let obj:SuperFieldAccess = <SuperFieldAccess>theEObject;
				let result:T = this.caseSuperFieldAccess(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.caseAbstractTypeQualifiedExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.SUPERMETHODINVOCATION: {
				let obj:SuperMethodInvocation = <SuperMethodInvocation>theEObject;
				let result:T = this.caseSuperMethodInvocation(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.caseAbstractTypeQualifiedExpression(obj);
				if (result == null) result = this.caseAbstractMethodInvocation(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.SWITCHCASE: {
				let obj:SwitchCase = <SwitchCase>theEObject;
				let result:T = this.caseSwitchCase(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.SWITCHSTATEMENT: {
				let obj:SwitchStatement = <SwitchStatement>theEObject;
				let result:T = this.caseSwitchStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.SYNCHRONIZEDSTATEMENT: {
				let obj:SynchronizedStatement = <SynchronizedStatement>theEObject;
				let result:T = this.caseSynchronizedStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.TAGELEMENT: {
				let obj:TagElement = <TagElement>theEObject;
				let result:T = this.caseTagElement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.TEXTELEMENT: {
				let obj:TextElement = <TextElement>theEObject;
				let result:T = this.caseTextElement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.THISEXPRESSION: {
				let obj:ThisExpression = <ThisExpression>theEObject;
				let result:T = this.caseThisExpression(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.caseAbstractTypeQualifiedExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.THROWSTATEMENT: {
				let obj:ThrowStatement = <ThrowStatement>theEObject;
				let result:T = this.caseThrowStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.TRYSTATEMENT: {
				let obj:TryStatement = <TryStatement>theEObject;
				let result:T = this.caseTryStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.TYPE: {
				let obj:Type = <Type>theEObject;
				let result:T = this.caseType(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.TYPEACCESS: {
				let obj:TypeAccess = <TypeAccess>theEObject;
				let result:T = this.caseTypeAccess(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.caseNamespaceAccess(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.TYPEDECLARATION: {
				let obj:TypeDeclaration = <TypeDeclaration>theEObject;
				let result:T = this.caseTypeDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.caseAbstractTypeDeclaration(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.TYPEDECLARATIONSTATEMENT: {
				let obj:TypeDeclarationStatement = <TypeDeclarationStatement>theEObject;
				let result:T = this.caseTypeDeclarationStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.TYPELITERAL: {
				let obj:TypeLiteral = <TypeLiteral>theEObject;
				let result:T = this.caseTypeLiteral(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.TYPEPARAMETER: {
				let obj:TypeParameter = <TypeParameter>theEObject;
				let result:T = this.caseTypeParameter(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.UNRESOLVEDITEM: {
				let obj:UnresolvedItem = <UnresolvedItem>theEObject;
				let result:T = this.caseUnresolvedItem(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.UNRESOLVEDITEMACCESS: {
				let obj:UnresolvedItemAccess = <UnresolvedItemAccess>theEObject;
				let result:T = this.caseUnresolvedItemAccess(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.caseNamespaceAccess(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.UNRESOLVEDANNOTATIONDECLARATION: {
				let obj:UnresolvedAnnotationDeclaration = <UnresolvedAnnotationDeclaration>theEObject;
				let result:T = this.caseUnresolvedAnnotationDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.caseAbstractTypeDeclaration(obj);
				if (result == null) result = this.caseAnnotationTypeDeclaration(obj);
				if (result == null) result = this.caseUnresolvedItem(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.UNRESOLVEDANNOTATIONTYPEMEMBERDECLARATION: {
				let obj:UnresolvedAnnotationTypeMemberDeclaration = <UnresolvedAnnotationTypeMemberDeclaration>theEObject;
				let result:T = this.caseUnresolvedAnnotationTypeMemberDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseAnnotationTypeMemberDeclaration(obj);
				if (result == null) result = this.caseUnresolvedItem(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.UNRESOLVEDCLASSDECLARATION: {
				let obj:UnresolvedClassDeclaration = <UnresolvedClassDeclaration>theEObject;
				let result:T = this.caseUnresolvedClassDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.caseAbstractTypeDeclaration(obj);
				if (result == null) result = this.caseTypeDeclaration(obj);
				if (result == null) result = this.caseClassDeclaration(obj);
				if (result == null) result = this.caseUnresolvedItem(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.UNRESOLVEDENUMDECLARATION: {
				let obj:UnresolvedEnumDeclaration = <UnresolvedEnumDeclaration>theEObject;
				let result:T = this.caseUnresolvedEnumDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.caseAbstractTypeDeclaration(obj);
				if (result == null) result = this.caseEnumDeclaration(obj);
				if (result == null) result = this.caseUnresolvedItem(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.UNRESOLVEDINTERFACEDECLARATION: {
				let obj:UnresolvedInterfaceDeclaration = <UnresolvedInterfaceDeclaration>theEObject;
				let result:T = this.caseUnresolvedInterfaceDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.caseAbstractTypeDeclaration(obj);
				if (result == null) result = this.caseTypeDeclaration(obj);
				if (result == null) result = this.caseInterfaceDeclaration(obj);
				if (result == null) result = this.caseUnresolvedItem(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.UNRESOLVEDLABELEDSTATEMENT: {
				let obj:UnresolvedLabeledStatement = <UnresolvedLabeledStatement>theEObject;
				let result:T = this.caseUnresolvedLabeledStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.caseLabeledStatement(obj);
				if (result == null) result = this.caseUnresolvedItem(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.UNRESOLVEDMETHODDECLARATION: {
				let obj:UnresolvedMethodDeclaration = <UnresolvedMethodDeclaration>theEObject;
				let result:T = this.caseUnresolvedMethodDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseAbstractMethodDeclaration(obj);
				if (result == null) result = this.caseMethodDeclaration(obj);
				if (result == null) result = this.caseUnresolvedItem(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.UNRESOLVEDSINGLEVARIABLEDECLARATION: {
				let obj:UnresolvedSingleVariableDeclaration = <UnresolvedSingleVariableDeclaration>theEObject;
				let result:T = this.caseUnresolvedSingleVariableDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseVariableDeclaration(obj);
				if (result == null) result = this.caseSingleVariableDeclaration(obj);
				if (result == null) result = this.caseUnresolvedItem(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.UNRESOLVEDTYPE: {
				let obj:UnresolvedType = <UnresolvedType>theEObject;
				let result:T = this.caseUnresolvedType(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.caseUnresolvedItem(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.UNRESOLVEDTYPEDECLARATION: {
				let obj:UnresolvedTypeDeclaration = <UnresolvedTypeDeclaration>theEObject;
				let result:T = this.caseUnresolvedTypeDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseBodyDeclaration(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.caseAbstractTypeDeclaration(obj);
				if (result == null) result = this.caseUnresolvedItem(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.UNRESOLVEDVARIABLEDECLARATIONFRAGMENT: {
				let obj:UnresolvedVariableDeclarationFragment = <UnresolvedVariableDeclarationFragment>theEObject;
				let result:T = this.caseUnresolvedVariableDeclarationFragment(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseVariableDeclaration(obj);
				if (result == null) result = this.caseVariableDeclarationFragment(obj);
				if (result == null) result = this.caseUnresolvedItem(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.VARIABLEDECLARATION: {
				let obj:VariableDeclaration = <VariableDeclaration>theEObject;
				let result:T = this.caseVariableDeclaration(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.VARIABLEDECLARATIONEXPRESSION: {
				let obj:VariableDeclarationExpression = <VariableDeclarationExpression>theEObject;
				let result:T = this.caseVariableDeclarationExpression(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseExpression(obj);
				if (result == null) result = this.caseAbstractVariablesContainer(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.VARIABLEDECLARATIONFRAGMENT: {
				let obj:VariableDeclarationFragment = <VariableDeclarationFragment>theEObject;
				let result:T = this.caseVariableDeclarationFragment(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseVariableDeclaration(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.VARIABLEDECLARATIONSTATEMENT: {
				let obj:VariableDeclarationStatement = <VariableDeclarationStatement>theEObject;
				let result:T = this.caseVariableDeclarationStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.caseAbstractVariablesContainer(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.WILDCARDTYPE: {
				let obj:WildCardType = <WildCardType>theEObject;
				let result:T = this.caseWildCardType(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseNamedElement(obj);
				if (result == null) result = this.caseType(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			case Java_PackageImpl.WHILESTATEMENT: {
				let obj:WhileStatement = <WhileStatement>theEObject;
				let result:T = this.caseWhileStatement(obj);
				if (result == null) result = this.caseASTNode(obj);
				if (result == null) result = this.caseStatement(obj);
				if (result == null) result = this.defaultCase(theEObject);
				return result;
			}
			default: return this.defaultCase(theEObject);
		}
	}
	
	
	public caseAbstractMethodDeclaration(object:AbstractMethodDeclaration):T {
		return null;
	}
	public caseAbstractMethodInvocation(object:AbstractMethodInvocation):T {
		return null;
	}
	public caseAbstractTypeDeclaration(object:AbstractTypeDeclaration):T {
		return null;
	}
	public caseAbstractTypeQualifiedExpression(object:AbstractTypeQualifiedExpression):T {
		return null;
	}
	public caseAbstractVariablesContainer(object:AbstractVariablesContainer):T {
		return null;
	}
	public caseAnnotation(object:Annotation):T {
		return null;
	}
	public caseArchive(object:Archive):T {
		return null;
	}
	public caseAssertStatement(object:AssertStatement):T {
		return null;
	}
	public caseASTNode(object:ASTNode):T {
		return null;
	}
	public caseAnnotationMemberValuePair(object:AnnotationMemberValuePair):T {
		return null;
	}
	public caseAnnotationTypeDeclaration(object:AnnotationTypeDeclaration):T {
		return null;
	}
	public caseAnnotationTypeMemberDeclaration(object:AnnotationTypeMemberDeclaration):T {
		return null;
	}
	public caseAnonymousClassDeclaration(object:AnonymousClassDeclaration):T {
		return null;
	}
	public caseArrayAccess(object:ArrayAccess):T {
		return null;
	}
	public caseArrayCreation(object:ArrayCreation):T {
		return null;
	}
	public caseArrayInitializer(object:ArrayInitializer):T {
		return null;
	}
	public caseArrayLengthAccess(object:ArrayLengthAccess):T {
		return null;
	}
	public caseArrayType(object:ArrayType):T {
		return null;
	}
	public caseAssignment(object:Assignment):T {
		return null;
	}
	public caseBodyDeclaration(object:BodyDeclaration):T {
		return null;
	}
	public caseBooleanLiteral(object:BooleanLiteral):T {
		return null;
	}
	public caseBlockComment(object:BlockComment):T {
		return null;
	}
	public caseBlock(object:Block):T {
		return null;
	}
	public caseBreakStatement(object:BreakStatement):T {
		return null;
	}
	public caseCastExpression(object:CastExpression):T {
		return null;
	}
	public caseCatchClause(object:CatchClause):T {
		return null;
	}
	public caseCharacterLiteral(object:CharacterLiteral):T {
		return null;
	}
	public caseClassFile(object:ClassFile):T {
		return null;
	}
	public caseClassInstanceCreation(object:ClassInstanceCreation):T {
		return null;
	}
	public caseConstructorDeclaration(object:ConstructorDeclaration):T {
		return null;
	}
	public caseConditionalExpression(object:ConditionalExpression):T {
		return null;
	}
	public caseConstructorInvocation(object:ConstructorInvocation):T {
		return null;
	}
	public caseClassDeclaration(object:ClassDeclaration):T {
		return null;
	}
	public caseComment(object:Comment):T {
		return null;
	}
	public caseCompilationUnit(object:CompilationUnit):T {
		return null;
	}
	public caseContinueStatement(object:ContinueStatement):T {
		return null;
	}
	public caseDoStatement(object:DoStatement):T {
		return null;
	}
	public caseEmptyStatement(object:EmptyStatement):T {
		return null;
	}
	public caseEnhancedForStatement(object:EnhancedForStatement):T {
		return null;
	}
	public caseEnumConstantDeclaration(object:EnumConstantDeclaration):T {
		return null;
	}
	public caseEnumDeclaration(object:EnumDeclaration):T {
		return null;
	}
	public caseExpression(object:Expression):T {
		return null;
	}
	public caseExpressionStatement(object:ExpressionStatement):T {
		return null;
	}
	public caseFieldAccess(object:FieldAccess):T {
		return null;
	}
	public caseFieldDeclaration(object:FieldDeclaration):T {
		return null;
	}
	public caseForStatement(object:ForStatement):T {
		return null;
	}
	public caseIfStatement(object:IfStatement):T {
		return null;
	}
	public caseImportDeclaration(object:ImportDeclaration):T {
		return null;
	}
	public caseInfixExpression(object:InfixExpression):T {
		return null;
	}
	public caseInitializer(object:Initializer):T {
		return null;
	}
	public caseInstanceofExpression(object:InstanceofExpression):T {
		return null;
	}
	public caseInterfaceDeclaration(object:InterfaceDeclaration):T {
		return null;
	}
	public caseJavadoc(object:Javadoc):T {
		return null;
	}
	public caseLabeledStatement(object:LabeledStatement):T {
		return null;
	}
	public caseLineComment(object:LineComment):T {
		return null;
	}
	public caseManifest(object:Manifest):T {
		return null;
	}
	public caseManifestAttribute(object:ManifestAttribute):T {
		return null;
	}
	public caseManifestEntry(object:ManifestEntry):T {
		return null;
	}
	public caseMemberRef(object:MemberRef):T {
		return null;
	}
	public caseMethodDeclaration(object:MethodDeclaration):T {
		return null;
	}
	public caseMethodInvocation(object:MethodInvocation):T {
		return null;
	}
	public caseMethodRef(object:MethodRef):T {
		return null;
	}
	public caseMethodRefParameter(object:MethodRefParameter):T {
		return null;
	}
	public caseModel(object:Model):T {
		return null;
	}
	public caseModifier(object:Modifier):T {
		return null;
	}
	public caseNamedElement(object:NamedElement):T {
		return null;
	}
	public caseNamespaceAccess(object:NamespaceAccess):T {
		return null;
	}
	public caseNumberLiteral(object:NumberLiteral):T {
		return null;
	}
	public caseNullLiteral(object:NullLiteral):T {
		return null;
	}
	public casePackage(object:Package):T {
		return null;
	}
	public casePackageAccess(object:PackageAccess):T {
		return null;
	}
	public caseParameterizedType(object:ParameterizedType):T {
		return null;
	}
	public caseParenthesizedExpression(object:ParenthesizedExpression):T {
		return null;
	}
	public casePostfixExpression(object:PostfixExpression):T {
		return null;
	}
	public casePrefixExpression(object:PrefixExpression):T {
		return null;
	}
	public casePrimitiveType(object:PrimitiveType):T {
		return null;
	}
	public casePrimitiveTypeBoolean(object:PrimitiveTypeBoolean):T {
		return null;
	}
	public casePrimitiveTypeByte(object:PrimitiveTypeByte):T {
		return null;
	}
	public casePrimitiveTypeChar(object:PrimitiveTypeChar):T {
		return null;
	}
	public casePrimitiveTypeDouble(object:PrimitiveTypeDouble):T {
		return null;
	}
	public casePrimitiveTypeShort(object:PrimitiveTypeShort):T {
		return null;
	}
	public casePrimitiveTypeFloat(object:PrimitiveTypeFloat):T {
		return null;
	}
	public casePrimitiveTypeInt(object:PrimitiveTypeInt):T {
		return null;
	}
	public casePrimitiveTypeLong(object:PrimitiveTypeLong):T {
		return null;
	}
	public casePrimitiveTypeVoid(object:PrimitiveTypeVoid):T {
		return null;
	}
	public caseReturnStatement(object:ReturnStatement):T {
		return null;
	}
	public caseSingleVariableAccess(object:SingleVariableAccess):T {
		return null;
	}
	public caseSingleVariableDeclaration(object:SingleVariableDeclaration):T {
		return null;
	}
	public caseStatement(object:Statement):T {
		return null;
	}
	public caseStringLiteral(object:StringLiteral):T {
		return null;
	}
	public caseSuperConstructorInvocation(object:SuperConstructorInvocation):T {
		return null;
	}
	public caseSuperFieldAccess(object:SuperFieldAccess):T {
		return null;
	}
	public caseSuperMethodInvocation(object:SuperMethodInvocation):T {
		return null;
	}
	public caseSwitchCase(object:SwitchCase):T {
		return null;
	}
	public caseSwitchStatement(object:SwitchStatement):T {
		return null;
	}
	public caseSynchronizedStatement(object:SynchronizedStatement):T {
		return null;
	}
	public caseTagElement(object:TagElement):T {
		return null;
	}
	public caseTextElement(object:TextElement):T {
		return null;
	}
	public caseThisExpression(object:ThisExpression):T {
		return null;
	}
	public caseThrowStatement(object:ThrowStatement):T {
		return null;
	}
	public caseTryStatement(object:TryStatement):T {
		return null;
	}
	public caseType(object:Type):T {
		return null;
	}
	public caseTypeAccess(object:TypeAccess):T {
		return null;
	}
	public caseTypeDeclaration(object:TypeDeclaration):T {
		return null;
	}
	public caseTypeDeclarationStatement(object:TypeDeclarationStatement):T {
		return null;
	}
	public caseTypeLiteral(object:TypeLiteral):T {
		return null;
	}
	public caseTypeParameter(object:TypeParameter):T {
		return null;
	}
	public caseUnresolvedItem(object:UnresolvedItem):T {
		return null;
	}
	public caseUnresolvedItemAccess(object:UnresolvedItemAccess):T {
		return null;
	}
	public caseUnresolvedAnnotationDeclaration(object:UnresolvedAnnotationDeclaration):T {
		return null;
	}
	public caseUnresolvedAnnotationTypeMemberDeclaration(object:UnresolvedAnnotationTypeMemberDeclaration):T {
		return null;
	}
	public caseUnresolvedClassDeclaration(object:UnresolvedClassDeclaration):T {
		return null;
	}
	public caseUnresolvedEnumDeclaration(object:UnresolvedEnumDeclaration):T {
		return null;
	}
	public caseUnresolvedInterfaceDeclaration(object:UnresolvedInterfaceDeclaration):T {
		return null;
	}
	public caseUnresolvedLabeledStatement(object:UnresolvedLabeledStatement):T {
		return null;
	}
	public caseUnresolvedMethodDeclaration(object:UnresolvedMethodDeclaration):T {
		return null;
	}
	public caseUnresolvedSingleVariableDeclaration(object:UnresolvedSingleVariableDeclaration):T {
		return null;
	}
	public caseUnresolvedType(object:UnresolvedType):T {
		return null;
	}
	public caseUnresolvedTypeDeclaration(object:UnresolvedTypeDeclaration):T {
		return null;
	}
	public caseUnresolvedVariableDeclarationFragment(object:UnresolvedVariableDeclarationFragment):T {
		return null;
	}
	public caseVariableDeclaration(object:VariableDeclaration):T {
		return null;
	}
	public caseVariableDeclarationExpression(object:VariableDeclarationExpression):T {
		return null;
	}
	public caseVariableDeclarationFragment(object:VariableDeclarationFragment):T {
		return null;
	}
	public caseVariableDeclarationStatement(object:VariableDeclarationStatement):T {
		return null;
	}
	public caseWildCardType(object:WildCardType):T {
		return null;
	}
	public caseWhileStatement(object:WhileStatement):T {
		return null;
	}
	
}
