// java_/Java_FactoryImpl.ts
import {ForStatement} from "java_/ForStatement";
import {MemberRef} from "java_/MemberRef";
import {CompilationUnit} from "java_/CompilationUnit";
import {SynchronizedStatementImpl} from "java_/SynchronizedStatementImpl";
import {ClassDeclarationImpl} from "java_/ClassDeclarationImpl";
import {VariableDeclarationFragment} from "java_/VariableDeclarationFragment";
import {ConstructorDeclaration} from "java_/ConstructorDeclaration";
import {ContinueStatement} from "java_/ContinueStatement";
import {ArrayAccessImpl} from "java_/ArrayAccessImpl";
import {AnnotationMemberValuePairImpl} from "java_/AnnotationMemberValuePairImpl";
import {PrimitiveTypeIntImpl} from "java_/PrimitiveTypeIntImpl";
import {SwitchCase} from "java_/SwitchCase";
import {ConstructorInvocationImpl} from "java_/ConstructorInvocationImpl";
import {DoStatementImpl} from "java_/DoStatementImpl";
import {Statement} from "java_/Statement";
import {ForStatementImpl} from "java_/ForStatementImpl";
import {ClassFile} from "java_/ClassFile";
import {TypeParameter} from "java_/TypeParameter";
import {EmptyStatementImpl} from "java_/EmptyStatementImpl";
import {Package} from "java_/Package";
import {WildCardTypeImpl} from "java_/WildCardTypeImpl";
import {NamespaceAccessImpl} from "java_/NamespaceAccessImpl";
import {AssignmentImpl} from "java_/AssignmentImpl";
import {TagElementImpl} from "java_/TagElementImpl";
import {PackageImpl} from "java_/PackageImpl";
import {MethodRefParameterImpl} from "java_/MethodRefParameterImpl";
import {PrimitiveTypeLong} from "java_/PrimitiveTypeLong";
import {TypeDeclarationStatementImpl} from "java_/TypeDeclarationStatementImpl";
import {FieldDeclarationImpl} from "java_/FieldDeclarationImpl";
import {ArrayLengthAccess} from "java_/ArrayLengthAccess";
import {MethodDeclaration} from "java_/MethodDeclaration";
import {ParenthesizedExpressionImpl} from "java_/ParenthesizedExpressionImpl";
import {ManifestEntry} from "java_/ManifestEntry";
import {NullLiteral} from "java_/NullLiteral";
import {ParameterizedType} from "java_/ParameterizedType";
import {LabeledStatement} from "java_/LabeledStatement";
import {InterfaceDeclaration} from "java_/InterfaceDeclaration";
import {InstanceofExpressionImpl} from "java_/InstanceofExpressionImpl";
import {TextElementImpl} from "java_/TextElementImpl";
import {ClassInstanceCreation} from "java_/ClassInstanceCreation";
import {PackageAccessImpl} from "java_/PackageAccessImpl";
import {UnresolvedTypeDeclarationImpl} from "java_/UnresolvedTypeDeclarationImpl";
import {CastExpressionImpl} from "java_/CastExpressionImpl";
import {ArchiveImpl} from "java_/ArchiveImpl";
import {UnresolvedLabeledStatement} from "java_/UnresolvedLabeledStatement";
import {UnresolvedInterfaceDeclarationImpl} from "java_/UnresolvedInterfaceDeclarationImpl";
import {FieldAccess} from "java_/FieldAccess";
import {BodyDeclaration} from "java_/BodyDeclaration";
import {Javadoc} from "java_/Javadoc";
import {UnresolvedEnumDeclaration} from "java_/UnresolvedEnumDeclaration";
import {InfixExpressionKind} from "java_/InfixExpressionKind";
import {IfStatementImpl} from "java_/IfStatementImpl";
import {VisibilityKind} from "java_/VisibilityKind";
import {AbstractMethodDeclaration} from "java_/AbstractMethodDeclaration";
import {EDataType} from "ecore/EDataType";
import {TypeDeclarationImpl} from "java_/TypeDeclarationImpl";
import {ImportDeclaration} from "java_/ImportDeclaration";
import {PrimitiveTypeFloat} from "java_/PrimitiveTypeFloat";
import {VariableDeclarationStatement} from "java_/VariableDeclarationStatement";
import {FieldAccessImpl} from "java_/FieldAccessImpl";
import {Model} from "java_/Model";
import {MemberRefImpl} from "java_/MemberRefImpl";
import {AbstractTypeQualifiedExpression} from "java_/AbstractTypeQualifiedExpression";
import {AnonymousClassDeclaration} from "java_/AnonymousClassDeclaration";
import {IfStatement} from "java_/IfStatement";
import {UnresolvedAnnotationDeclaration} from "java_/UnresolvedAnnotationDeclaration";
import {ExpressionImpl} from "java_/ExpressionImpl";
import {EClass} from "ecore/EClass";
import {UnresolvedEnumDeclarationImpl} from "java_/UnresolvedEnumDeclarationImpl";
import {BreakStatement} from "java_/BreakStatement";
import {UnresolvedItem} from "java_/UnresolvedItem";
import {BlockComment} from "java_/BlockComment";
import {CatchClause} from "java_/CatchClause";
import {UnresolvedSingleVariableDeclaration} from "java_/UnresolvedSingleVariableDeclaration";
import {NumberLiteralImpl} from "java_/NumberLiteralImpl";
import {PrimitiveTypeImpl} from "java_/PrimitiveTypeImpl";
import {CastExpression} from "java_/CastExpression";
import {Expression} from "java_/Expression";
import {VariableDeclarationStatementImpl} from "java_/VariableDeclarationStatementImpl";
import {ParenthesizedExpression} from "java_/ParenthesizedExpression";
import {SuperFieldAccess} from "java_/SuperFieldAccess";
import {UnresolvedItemImpl} from "java_/UnresolvedItemImpl";
import {ExpressionStatementImpl} from "java_/ExpressionStatementImpl";
import {UnresolvedSingleVariableDeclarationImpl} from "java_/UnresolvedSingleVariableDeclarationImpl";
import {AbstractTypeDeclarationImpl} from "java_/AbstractTypeDeclarationImpl";
import {SingleVariableAccess} from "java_/SingleVariableAccess";
import {PostfixExpressionKind} from "java_/PostfixExpressionKind";
import {SingleVariableDeclarationImpl} from "java_/SingleVariableDeclarationImpl";
import {AnnotationTypeDeclaration} from "java_/AnnotationTypeDeclaration";
import {DoStatement} from "java_/DoStatement";
import {TypeImpl} from "java_/TypeImpl";
import {EnhancedForStatementImpl} from "java_/EnhancedForStatementImpl";
import {EnumDeclaration} from "java_/EnumDeclaration";
import {PrimitiveTypeByteImpl} from "java_/PrimitiveTypeByteImpl";
import {SuperMethodInvocation} from "java_/SuperMethodInvocation";
import {StringLiteralImpl} from "java_/StringLiteralImpl";
import {UnresolvedLabeledStatementImpl} from "java_/UnresolvedLabeledStatementImpl";
import {ModifierImpl} from "java_/ModifierImpl";
import {AssertStatement} from "java_/AssertStatement";
import {NamedElementImpl} from "java_/NamedElementImpl";
import {FieldDeclaration} from "java_/FieldDeclaration";
import {PackageAccess} from "java_/PackageAccess";
import {VariableDeclaration} from "java_/VariableDeclaration";
import {BodyDeclarationImpl} from "java_/BodyDeclarationImpl";
import {CatchClauseImpl} from "java_/CatchClauseImpl";
import {ArrayInitializerImpl} from "java_/ArrayInitializerImpl";
import {UnresolvedMethodDeclaration} from "java_/UnresolvedMethodDeclaration";
import {Manifest} from "java_/Manifest";
import {SwitchStatement} from "java_/SwitchStatement";
import {AnnotationTypeDeclarationImpl} from "java_/AnnotationTypeDeclarationImpl";
import {AnonymousClassDeclarationImpl} from "java_/AnonymousClassDeclarationImpl";
import {ArrayInitializer} from "java_/ArrayInitializer";
import {ManifestImpl} from "java_/ManifestImpl";
import {Assignment} from "java_/Assignment";
import {AnnotationTypeMemberDeclaration} from "java_/AnnotationTypeMemberDeclaration";
import {BooleanLiteralImpl} from "java_/BooleanLiteralImpl";
import {PrimitiveTypeBoolean} from "java_/PrimitiveTypeBoolean";
import {EmptyStatement} from "java_/EmptyStatement";
import {InterfaceDeclarationImpl} from "java_/InterfaceDeclarationImpl";
import {MethodInvocationImpl} from "java_/MethodInvocationImpl";
import {PrefixExpressionImpl} from "java_/PrefixExpressionImpl";
import {UnresolvedItemAccessImpl} from "java_/UnresolvedItemAccessImpl";
import {UnresolvedType} from "java_/UnresolvedType";
import {VariableDeclarationFragmentImpl} from "java_/VariableDeclarationFragmentImpl";
import {MethodRefImpl} from "java_/MethodRefImpl";
import {AbstractVariablesContainerImpl} from "java_/AbstractVariablesContainerImpl";
import {PrimitiveTypeShortImpl} from "java_/PrimitiveTypeShortImpl";
import {NullLiteralImpl} from "java_/NullLiteralImpl";
import {AnnotationMemberValuePair} from "java_/AnnotationMemberValuePair";
import {InstanceofExpression} from "java_/InstanceofExpression";
import {ClassFileImpl} from "java_/ClassFileImpl";
import {SingleVariableAccessImpl} from "java_/SingleVariableAccessImpl";
import {AbstractMethodDeclarationImpl} from "java_/AbstractMethodDeclarationImpl";
import {PrimitiveTypeShort} from "java_/PrimitiveTypeShort";
import {WildCardType} from "java_/WildCardType";
import {MethodInvocation} from "java_/MethodInvocation";
import {SuperConstructorInvocation} from "java_/SuperConstructorInvocation";
import {BlockCommentImpl} from "java_/BlockCommentImpl";
import {ReturnStatementImpl} from "java_/ReturnStatementImpl";
import {MethodDeclarationImpl} from "java_/MethodDeclarationImpl";
import {UnresolvedAnnotationDeclarationImpl} from "java_/UnresolvedAnnotationDeclarationImpl";
import {AnnotationImpl} from "java_/AnnotationImpl";
import {AbstractMethodInvocation} from "java_/AbstractMethodInvocation";
import {ParameterizedTypeImpl} from "java_/ParameterizedTypeImpl";
import {AbstractTypeDeclaration} from "java_/AbstractTypeDeclaration";
import {TextElement} from "java_/TextElement";
import {ConstructorDeclarationImpl} from "java_/ConstructorDeclarationImpl";
import {ManifestAttributeImpl} from "java_/ManifestAttributeImpl";
import {PostfixExpressionImpl} from "java_/PostfixExpressionImpl";
import {TypeDeclaration} from "java_/TypeDeclaration";
import {TypeParameterImpl} from "java_/TypeParameterImpl";
import {ASTNodeImpl} from "java_/ASTNodeImpl";
import {ModelImpl} from "java_/ModelImpl";
import {PrimitiveTypeVoid} from "java_/PrimitiveTypeVoid";
import {ThrowStatement} from "java_/ThrowStatement";
import {PrimitiveTypeBooleanImpl} from "java_/PrimitiveTypeBooleanImpl";
import {ThisExpression} from "java_/ThisExpression";
import {PostfixExpression} from "java_/PostfixExpression";
import {AbstractVariablesContainer} from "java_/AbstractVariablesContainer";
import {EnumConstantDeclaration} from "java_/EnumConstantDeclaration";
import {Archive} from "java_/Archive";
import {LineComment} from "java_/LineComment";
import {UnresolvedClassDeclaration} from "java_/UnresolvedClassDeclaration";
import {Java_PackageImpl} from "java_/Java_PackageImpl";
import {LabeledStatementImpl} from "java_/LabeledStatementImpl";
import {SuperMethodInvocationImpl} from "java_/SuperMethodInvocationImpl";
import {BlockImpl} from "java_/BlockImpl";
import {UnresolvedTypeImpl} from "java_/UnresolvedTypeImpl";
import {SingleVariableDeclaration} from "java_/SingleVariableDeclaration";
import {ThisExpressionImpl} from "java_/ThisExpressionImpl";
import {Modifier} from "java_/Modifier";
import {UnresolvedAnnotationTypeMemberDeclaration} from "java_/UnresolvedAnnotationTypeMemberDeclaration";
import {VariableDeclarationExpression} from "java_/VariableDeclarationExpression";
import {CompilationUnitImpl} from "java_/CompilationUnitImpl";
import {LineCommentImpl} from "java_/LineCommentImpl";
import {PrimitiveTypeDouble} from "java_/PrimitiveTypeDouble";
import {Annotation} from "java_/Annotation";
import {EObject} from "ecore/EObject";
import {ReturnStatement} from "java_/ReturnStatement";
import {CharacterLiteral} from "java_/CharacterLiteral";
import {EnhancedForStatement} from "java_/EnhancedForStatement";
import {ArrayLengthAccessImpl} from "java_/ArrayLengthAccessImpl";
import {AbstractTypeQualifiedExpressionImpl} from "java_/AbstractTypeQualifiedExpressionImpl";
import {NamedElement} from "java_/NamedElement";
import {PrimitiveTypeChar} from "java_/PrimitiveTypeChar";
import {ArrayCreation} from "java_/ArrayCreation";
import {SuperConstructorInvocationImpl} from "java_/SuperConstructorInvocationImpl";
import {TypeDeclarationStatement} from "java_/TypeDeclarationStatement";
import {ArrayType} from "java_/ArrayType";
import {EnumDeclarationImpl} from "java_/EnumDeclarationImpl";
import {UnresolvedInterfaceDeclaration} from "java_/UnresolvedInterfaceDeclaration";
import {NumberLiteral} from "java_/NumberLiteral";
import {PrimitiveTypeCharImpl} from "java_/PrimitiveTypeCharImpl";
import {PrefixExpressionKind} from "java_/PrefixExpressionKind";
import {AnnotationTypeMemberDeclarationImpl} from "java_/AnnotationTypeMemberDeclarationImpl";
import {ArrayCreationImpl} from "java_/ArrayCreationImpl";
import {ConditionalExpression} from "java_/ConditionalExpression";
import {Block} from "java_/Block";
import {BooleanLiteral} from "java_/BooleanLiteral";
import {SwitchCaseImpl} from "java_/SwitchCaseImpl";
import {UnresolvedClassDeclarationImpl} from "java_/UnresolvedClassDeclarationImpl";
import {EnumConstantDeclarationImpl} from "java_/EnumConstantDeclarationImpl";
import {TypeAccessImpl} from "java_/TypeAccessImpl";
import {VariableDeclarationImpl} from "java_/VariableDeclarationImpl";
import {PrefixExpression} from "java_/PrefixExpression";
import {PrimitiveTypeLongImpl} from "java_/PrimitiveTypeLongImpl";
import {SwitchStatementImpl} from "java_/SwitchStatementImpl";
import {CharacterLiteralImpl} from "java_/CharacterLiteralImpl";
import {ThrowStatementImpl} from "java_/ThrowStatementImpl";
import {ClassDeclaration} from "java_/ClassDeclaration";
import {Type} from "java_/Type";
import {ASTNode} from "java_/ASTNode";
import {PrimitiveTypeVoidImpl} from "java_/PrimitiveTypeVoidImpl";
import {ArrayAccess} from "java_/ArrayAccess";
import {TypeAccess} from "java_/TypeAccess";
import {AssertStatementImpl} from "java_/AssertStatementImpl";
import {StatementImpl} from "java_/StatementImpl";
import {WhileStatement} from "java_/WhileStatement";
import {InheritanceKind} from "java_/InheritanceKind";
import {TypeLiteralImpl} from "java_/TypeLiteralImpl";
import {NamespaceAccess} from "java_/NamespaceAccess";
import {CommentImpl} from "java_/CommentImpl";
import {TryStatementImpl} from "java_/TryStatementImpl";
import {ImportDeclarationImpl} from "java_/ImportDeclarationImpl";
import {ConditionalExpressionImpl} from "java_/ConditionalExpressionImpl";
import {ContinueStatementImpl} from "java_/ContinueStatementImpl";
import {ConstructorInvocation} from "java_/ConstructorInvocation";
import {AssignmentKind} from "java_/AssignmentKind";
import {ArrayTypeImpl} from "java_/ArrayTypeImpl";
import {MethodRef} from "java_/MethodRef";
import {UnresolvedVariableDeclarationFragmentImpl} from "java_/UnresolvedVariableDeclarationFragmentImpl";
import {PrimitiveTypeDoubleImpl} from "java_/PrimitiveTypeDoubleImpl";
import {BreakStatementImpl} from "java_/BreakStatementImpl";
import {StringLiteral} from "java_/StringLiteral";
import {ManifestEntryImpl} from "java_/ManifestEntryImpl";
import {Java_Factory} from "java_/Java_Factory";
import {UnresolvedTypeDeclaration} from "java_/UnresolvedTypeDeclaration";
import {Comment} from "java_/Comment";
import {WhileStatementImpl} from "java_/WhileStatementImpl";
import {TypeLiteral} from "java_/TypeLiteral";
import {InitializerImpl} from "java_/InitializerImpl";
import {Initializer} from "java_/Initializer";
import {PrimitiveType} from "java_/PrimitiveType";
import {UnresolvedMethodDeclarationImpl} from "java_/UnresolvedMethodDeclarationImpl";
import {UnresolvedVariableDeclarationFragment} from "java_/UnresolvedVariableDeclarationFragment";
import {ManifestAttribute} from "java_/ManifestAttribute";
import {PrimitiveTypeFloatImpl} from "java_/PrimitiveTypeFloatImpl";
import {EFactoryImpl} from "ecore/EFactoryImpl";
import {ClassInstanceCreationImpl} from "java_/ClassInstanceCreationImpl";
import {PrimitiveTypeInt} from "java_/PrimitiveTypeInt";
import {SynchronizedStatement} from "java_/SynchronizedStatement";
import {JavadocImpl} from "java_/JavadocImpl";
import {PrimitiveTypeByte} from "java_/PrimitiveTypeByte";
import {UnresolvedAnnotationTypeMemberDeclarationImpl} from "java_/UnresolvedAnnotationTypeMemberDeclarationImpl";
import {UnresolvedItemAccess} from "java_/UnresolvedItemAccess";
import {SuperFieldAccessImpl} from "java_/SuperFieldAccessImpl";
import {MethodRefParameter} from "java_/MethodRefParameter";
import {TryStatement} from "java_/TryStatement";
import {AbstractMethodInvocationImpl} from "java_/AbstractMethodInvocationImpl";
import {InfixExpression} from "java_/InfixExpression";
import {InfixExpressionImpl} from "java_/InfixExpressionImpl";
import {VariableDeclarationExpressionImpl} from "java_/VariableDeclarationExpressionImpl";
import {TagElement} from "java_/TagElement";
import {ExpressionStatement} from "java_/ExpressionStatement";
export class Java_FactoryImpl extends EFactoryImpl implements Java_Factory{
	public static eINSTANCE : Java_Factory = Java_FactoryImpl.init();
	public static init() : Java_Factory 
	{
		return new Java_FactoryImpl();
	}
	
	public createAbstractMethodDeclaration = () : AbstractMethodDeclaration => {
		var theAbstractMethodDeclaration = new AbstractMethodDeclarationImpl();
		
		
		AbstractMethodDeclarationImpl.allInstances_.add(theAbstractMethodDeclaration);
		
		return theAbstractMethodDeclaration;
	}
	public createAbstractMethodInvocation = () : AbstractMethodInvocation => {
		var theAbstractMethodInvocation = new AbstractMethodInvocationImpl();
		
		
		AbstractMethodInvocationImpl.allInstances_.add(theAbstractMethodInvocation);
		
		return theAbstractMethodInvocation;
	}
	public createAbstractTypeDeclaration = () : AbstractTypeDeclaration => {
		var theAbstractTypeDeclaration = new AbstractTypeDeclarationImpl();
		
		
		AbstractTypeDeclarationImpl.allInstances_.add(theAbstractTypeDeclaration);
		
		return theAbstractTypeDeclaration;
	}
	public createAbstractTypeQualifiedExpression = () : AbstractTypeQualifiedExpression => {
		var theAbstractTypeQualifiedExpression = new AbstractTypeQualifiedExpressionImpl();
		
		
		AbstractTypeQualifiedExpressionImpl.allInstances_.add(theAbstractTypeQualifiedExpression);
		
		return theAbstractTypeQualifiedExpression;
	}
	public createAbstractVariablesContainer = () : AbstractVariablesContainer => {
		var theAbstractVariablesContainer = new AbstractVariablesContainerImpl();
		
		
		AbstractVariablesContainerImpl.allInstances_.add(theAbstractVariablesContainer);
		
		return theAbstractVariablesContainer;
	}
	public createAnnotation = () : Annotation => {
		var theAnnotation = new AnnotationImpl();
		
		
		AnnotationImpl.allInstances_.add(theAnnotation);
		
		return theAnnotation;
	}
	public createArchive = () : Archive => {
		var theArchive = new ArchiveImpl();
		
		
		ArchiveImpl.allInstances_.add(theArchive);
		
		return theArchive;
	}
	public createAssertStatement = () : AssertStatement => {
		var theAssertStatement = new AssertStatementImpl();
		
		
		AssertStatementImpl.allInstances_.add(theAssertStatement);
		
		return theAssertStatement;
	}
	public createASTNode = () : ASTNode => {
		var theASTNode = new ASTNodeImpl();
		
		
		ASTNodeImpl.allInstances_.add(theASTNode);
		
		return theASTNode;
	}
	public createAnnotationMemberValuePair = () : AnnotationMemberValuePair => {
		var theAnnotationMemberValuePair = new AnnotationMemberValuePairImpl();
		
		
		AnnotationMemberValuePairImpl.allInstances_.add(theAnnotationMemberValuePair);
		
		return theAnnotationMemberValuePair;
	}
	public createAnnotationTypeDeclaration = () : AnnotationTypeDeclaration => {
		var theAnnotationTypeDeclaration = new AnnotationTypeDeclarationImpl();
		
		
		AnnotationTypeDeclarationImpl.allInstances_.add(theAnnotationTypeDeclaration);
		
		return theAnnotationTypeDeclaration;
	}
	public createAnnotationTypeMemberDeclaration = () : AnnotationTypeMemberDeclaration => {
		var theAnnotationTypeMemberDeclaration = new AnnotationTypeMemberDeclarationImpl();
		
		
		AnnotationTypeMemberDeclarationImpl.allInstances_.add(theAnnotationTypeMemberDeclaration);
		
		return theAnnotationTypeMemberDeclaration;
	}
	public createAnonymousClassDeclaration = () : AnonymousClassDeclaration => {
		var theAnonymousClassDeclaration = new AnonymousClassDeclarationImpl();
		
		
		AnonymousClassDeclarationImpl.allInstances_.add(theAnonymousClassDeclaration);
		
		return theAnonymousClassDeclaration;
	}
	public createArrayAccess = () : ArrayAccess => {
		var theArrayAccess = new ArrayAccessImpl();
		
		
		ArrayAccessImpl.allInstances_.add(theArrayAccess);
		
		return theArrayAccess;
	}
	public createArrayCreation = () : ArrayCreation => {
		var theArrayCreation = new ArrayCreationImpl();
		
		
		ArrayCreationImpl.allInstances_.add(theArrayCreation);
		
		return theArrayCreation;
	}
	public createArrayInitializer = () : ArrayInitializer => {
		var theArrayInitializer = new ArrayInitializerImpl();
		
		
		ArrayInitializerImpl.allInstances_.add(theArrayInitializer);
		
		return theArrayInitializer;
	}
	public createArrayLengthAccess = () : ArrayLengthAccess => {
		var theArrayLengthAccess = new ArrayLengthAccessImpl();
		
		
		ArrayLengthAccessImpl.allInstances_.add(theArrayLengthAccess);
		
		return theArrayLengthAccess;
	}
	public createArrayType = () : ArrayType => {
		var theArrayType = new ArrayTypeImpl();
		
		
		ArrayTypeImpl.allInstances_.add(theArrayType);
		
		return theArrayType;
	}
	public createAssignment = () : Assignment => {
		var theAssignment = new AssignmentImpl();
		
		
		AssignmentImpl.allInstances_.add(theAssignment);
		
		return theAssignment;
	}
	public createBodyDeclaration = () : BodyDeclaration => {
		var theBodyDeclaration = new BodyDeclarationImpl();
		
		
		BodyDeclarationImpl.allInstances_.add(theBodyDeclaration);
		
		return theBodyDeclaration;
	}
	public createBooleanLiteral = () : BooleanLiteral => {
		var theBooleanLiteral = new BooleanLiteralImpl();
		
		
		BooleanLiteralImpl.allInstances_.add(theBooleanLiteral);
		
		return theBooleanLiteral;
	}
	public createBlockComment = () : BlockComment => {
		var theBlockComment = new BlockCommentImpl();
		
		
		BlockCommentImpl.allInstances_.add(theBlockComment);
		
		return theBlockComment;
	}
	public createBlock = () : Block => {
		var theBlock = new BlockImpl();
		
		
		BlockImpl.allInstances_.add(theBlock);
		
		return theBlock;
	}
	public createBreakStatement = () : BreakStatement => {
		var theBreakStatement = new BreakStatementImpl();
		
		
		BreakStatementImpl.allInstances_.add(theBreakStatement);
		
		return theBreakStatement;
	}
	public createCastExpression = () : CastExpression => {
		var theCastExpression = new CastExpressionImpl();
		
		
		CastExpressionImpl.allInstances_.add(theCastExpression);
		
		return theCastExpression;
	}
	public createCatchClause = () : CatchClause => {
		var theCatchClause = new CatchClauseImpl();
		
		
		CatchClauseImpl.allInstances_.add(theCatchClause);
		
		return theCatchClause;
	}
	public createCharacterLiteral = () : CharacterLiteral => {
		var theCharacterLiteral = new CharacterLiteralImpl();
		
		
		CharacterLiteralImpl.allInstances_.add(theCharacterLiteral);
		
		return theCharacterLiteral;
	}
	public createClassFile = () : ClassFile => {
		var theClassFile = new ClassFileImpl();
		
		
		ClassFileImpl.allInstances_.add(theClassFile);
		
		return theClassFile;
	}
	public createClassInstanceCreation = () : ClassInstanceCreation => {
		var theClassInstanceCreation = new ClassInstanceCreationImpl();
		
		
		ClassInstanceCreationImpl.allInstances_.add(theClassInstanceCreation);
		
		return theClassInstanceCreation;
	}
	public createConstructorDeclaration = () : ConstructorDeclaration => {
		var theConstructorDeclaration = new ConstructorDeclarationImpl();
		
		
		ConstructorDeclarationImpl.allInstances_.add(theConstructorDeclaration);
		
		return theConstructorDeclaration;
	}
	public createConditionalExpression = () : ConditionalExpression => {
		var theConditionalExpression = new ConditionalExpressionImpl();
		
		
		ConditionalExpressionImpl.allInstances_.add(theConditionalExpression);
		
		return theConditionalExpression;
	}
	public createConstructorInvocation = () : ConstructorInvocation => {
		var theConstructorInvocation = new ConstructorInvocationImpl();
		
		
		ConstructorInvocationImpl.allInstances_.add(theConstructorInvocation);
		
		return theConstructorInvocation;
	}
	public createClassDeclaration = () : ClassDeclaration => {
		var theClassDeclaration = new ClassDeclarationImpl();
		
		
		ClassDeclarationImpl.allInstances_.add(theClassDeclaration);
		
		return theClassDeclaration;
	}
	public createComment = () : Comment => {
		var theComment = new CommentImpl();
		
		
		CommentImpl.allInstances_.add(theComment);
		
		return theComment;
	}
	public createCompilationUnit = () : CompilationUnit => {
		var theCompilationUnit = new CompilationUnitImpl();
		
		
		CompilationUnitImpl.allInstances_.add(theCompilationUnit);
		
		return theCompilationUnit;
	}
	public createContinueStatement = () : ContinueStatement => {
		var theContinueStatement = new ContinueStatementImpl();
		
		
		ContinueStatementImpl.allInstances_.add(theContinueStatement);
		
		return theContinueStatement;
	}
	public createDoStatement = () : DoStatement => {
		var theDoStatement = new DoStatementImpl();
		
		
		DoStatementImpl.allInstances_.add(theDoStatement);
		
		return theDoStatement;
	}
	public createEmptyStatement = () : EmptyStatement => {
		var theEmptyStatement = new EmptyStatementImpl();
		
		
		EmptyStatementImpl.allInstances_.add(theEmptyStatement);
		
		return theEmptyStatement;
	}
	public createEnhancedForStatement = () : EnhancedForStatement => {
		var theEnhancedForStatement = new EnhancedForStatementImpl();
		
		
		EnhancedForStatementImpl.allInstances_.add(theEnhancedForStatement);
		
		return theEnhancedForStatement;
	}
	public createEnumConstantDeclaration = () : EnumConstantDeclaration => {
		var theEnumConstantDeclaration = new EnumConstantDeclarationImpl();
		
		
		EnumConstantDeclarationImpl.allInstances_.add(theEnumConstantDeclaration);
		
		return theEnumConstantDeclaration;
	}
	public createEnumDeclaration = () : EnumDeclaration => {
		var theEnumDeclaration = new EnumDeclarationImpl();
		
		
		EnumDeclarationImpl.allInstances_.add(theEnumDeclaration);
		
		return theEnumDeclaration;
	}
	public createExpression = () : Expression => {
		var theExpression = new ExpressionImpl();
		
		
		ExpressionImpl.allInstances_.add(theExpression);
		
		return theExpression;
	}
	public createExpressionStatement = () : ExpressionStatement => {
		var theExpressionStatement = new ExpressionStatementImpl();
		
		
		ExpressionStatementImpl.allInstances_.add(theExpressionStatement);
		
		return theExpressionStatement;
	}
	public createFieldAccess = () : FieldAccess => {
		var theFieldAccess = new FieldAccessImpl();
		
		
		FieldAccessImpl.allInstances_.add(theFieldAccess);
		
		return theFieldAccess;
	}
	public createFieldDeclaration = () : FieldDeclaration => {
		var theFieldDeclaration = new FieldDeclarationImpl();
		
		
		FieldDeclarationImpl.allInstances_.add(theFieldDeclaration);
		
		return theFieldDeclaration;
	}
	public createForStatement = () : ForStatement => {
		var theForStatement = new ForStatementImpl();
		
		
		ForStatementImpl.allInstances_.add(theForStatement);
		
		return theForStatement;
	}
	public createIfStatement = () : IfStatement => {
		var theIfStatement = new IfStatementImpl();
		
		
		IfStatementImpl.allInstances_.add(theIfStatement);
		
		return theIfStatement;
	}
	public createImportDeclaration = () : ImportDeclaration => {
		var theImportDeclaration = new ImportDeclarationImpl();
		
		
		ImportDeclarationImpl.allInstances_.add(theImportDeclaration);
		
		return theImportDeclaration;
	}
	public createInfixExpression = () : InfixExpression => {
		var theInfixExpression = new InfixExpressionImpl();
		
		
		InfixExpressionImpl.allInstances_.add(theInfixExpression);
		
		return theInfixExpression;
	}
	public createInitializer = () : Initializer => {
		var theInitializer = new InitializerImpl();
		
		
		InitializerImpl.allInstances_.add(theInitializer);
		
		return theInitializer;
	}
	public createInstanceofExpression = () : InstanceofExpression => {
		var theInstanceofExpression = new InstanceofExpressionImpl();
		
		
		InstanceofExpressionImpl.allInstances_.add(theInstanceofExpression);
		
		return theInstanceofExpression;
	}
	public createInterfaceDeclaration = () : InterfaceDeclaration => {
		var theInterfaceDeclaration = new InterfaceDeclarationImpl();
		
		
		InterfaceDeclarationImpl.allInstances_.add(theInterfaceDeclaration);
		
		return theInterfaceDeclaration;
	}
	public createJavadoc = () : Javadoc => {
		var theJavadoc = new JavadocImpl();
		
		
		JavadocImpl.allInstances_.add(theJavadoc);
		
		return theJavadoc;
	}
	public createLabeledStatement = () : LabeledStatement => {
		var theLabeledStatement = new LabeledStatementImpl();
		
		
		LabeledStatementImpl.allInstances_.add(theLabeledStatement);
		
		return theLabeledStatement;
	}
	public createLineComment = () : LineComment => {
		var theLineComment = new LineCommentImpl();
		
		
		LineCommentImpl.allInstances_.add(theLineComment);
		
		return theLineComment;
	}
	public createManifest = () : Manifest => {
		var theManifest = new ManifestImpl();
		
		
		ManifestImpl.allInstances_.add(theManifest);
		
		return theManifest;
	}
	public createManifestAttribute = () : ManifestAttribute => {
		var theManifestAttribute = new ManifestAttributeImpl();
		
		
		ManifestAttributeImpl.allInstances_.add(theManifestAttribute);
		
		return theManifestAttribute;
	}
	public createManifestEntry = () : ManifestEntry => {
		var theManifestEntry = new ManifestEntryImpl();
		
		
		ManifestEntryImpl.allInstances_.add(theManifestEntry);
		
		return theManifestEntry;
	}
	public createMemberRef = () : MemberRef => {
		var theMemberRef = new MemberRefImpl();
		
		
		MemberRefImpl.allInstances_.add(theMemberRef);
		
		return theMemberRef;
	}
	public createMethodDeclaration = () : MethodDeclaration => {
		var theMethodDeclaration = new MethodDeclarationImpl();
		
		
		MethodDeclarationImpl.allInstances_.add(theMethodDeclaration);
		
		return theMethodDeclaration;
	}
	public createMethodInvocation = () : MethodInvocation => {
		var theMethodInvocation = new MethodInvocationImpl();
		
		
		MethodInvocationImpl.allInstances_.add(theMethodInvocation);
		
		return theMethodInvocation;
	}
	public createMethodRef = () : MethodRef => {
		var theMethodRef = new MethodRefImpl();
		
		
		MethodRefImpl.allInstances_.add(theMethodRef);
		
		return theMethodRef;
	}
	public createMethodRefParameter = () : MethodRefParameter => {
		var theMethodRefParameter = new MethodRefParameterImpl();
		
		
		MethodRefParameterImpl.allInstances_.add(theMethodRefParameter);
		
		return theMethodRefParameter;
	}
	public createModel = () : Model => {
		var theModel = new ModelImpl();
		
		
		ModelImpl.allInstances_.add(theModel);
		
		return theModel;
	}
	public createModifier = () : Modifier => {
		var theModifier = new ModifierImpl();
		
		
		ModifierImpl.allInstances_.add(theModifier);
		
		return theModifier;
	}
	public createNamedElement = () : NamedElement => {
		var theNamedElement = new NamedElementImpl();
		
		
		NamedElementImpl.allInstances_.add(theNamedElement);
		
		return theNamedElement;
	}
	public createNamespaceAccess = () : NamespaceAccess => {
		var theNamespaceAccess = new NamespaceAccessImpl();
		
		
		NamespaceAccessImpl.allInstances_.add(theNamespaceAccess);
		
		return theNamespaceAccess;
	}
	public createNumberLiteral = () : NumberLiteral => {
		var theNumberLiteral = new NumberLiteralImpl();
		
		
		NumberLiteralImpl.allInstances_.add(theNumberLiteral);
		
		return theNumberLiteral;
	}
	public createNullLiteral = () : NullLiteral => {
		var theNullLiteral = new NullLiteralImpl();
		
		
		NullLiteralImpl.allInstances_.add(theNullLiteral);
		
		return theNullLiteral;
	}
	public createPackage = () : Package => {
		var thePackage = new PackageImpl();
		
		
		PackageImpl.allInstances_.add(thePackage);
		
		return thePackage;
	}
	public createPackageAccess = () : PackageAccess => {
		var thePackageAccess = new PackageAccessImpl();
		
		
		PackageAccessImpl.allInstances_.add(thePackageAccess);
		
		return thePackageAccess;
	}
	public createParameterizedType = () : ParameterizedType => {
		var theParameterizedType = new ParameterizedTypeImpl();
		
		
		ParameterizedTypeImpl.allInstances_.add(theParameterizedType);
		
		return theParameterizedType;
	}
	public createParenthesizedExpression = () : ParenthesizedExpression => {
		var theParenthesizedExpression = new ParenthesizedExpressionImpl();
		
		
		ParenthesizedExpressionImpl.allInstances_.add(theParenthesizedExpression);
		
		return theParenthesizedExpression;
	}
	public createPostfixExpression = () : PostfixExpression => {
		var thePostfixExpression = new PostfixExpressionImpl();
		
		
		PostfixExpressionImpl.allInstances_.add(thePostfixExpression);
		
		return thePostfixExpression;
	}
	public createPrefixExpression = () : PrefixExpression => {
		var thePrefixExpression = new PrefixExpressionImpl();
		
		
		PrefixExpressionImpl.allInstances_.add(thePrefixExpression);
		
		return thePrefixExpression;
	}
	public createPrimitiveType = () : PrimitiveType => {
		var thePrimitiveType = new PrimitiveTypeImpl();
		
		
		PrimitiveTypeImpl.allInstances_.add(thePrimitiveType);
		
		return thePrimitiveType;
	}
	public createPrimitiveTypeBoolean = () : PrimitiveTypeBoolean => {
		var thePrimitiveTypeBoolean = new PrimitiveTypeBooleanImpl();
		
		
		PrimitiveTypeBooleanImpl.allInstances_.add(thePrimitiveTypeBoolean);
		
		return thePrimitiveTypeBoolean;
	}
	public createPrimitiveTypeByte = () : PrimitiveTypeByte => {
		var thePrimitiveTypeByte = new PrimitiveTypeByteImpl();
		
		
		PrimitiveTypeByteImpl.allInstances_.add(thePrimitiveTypeByte);
		
		return thePrimitiveTypeByte;
	}
	public createPrimitiveTypeChar = () : PrimitiveTypeChar => {
		var thePrimitiveTypeChar = new PrimitiveTypeCharImpl();
		
		
		PrimitiveTypeCharImpl.allInstances_.add(thePrimitiveTypeChar);
		
		return thePrimitiveTypeChar;
	}
	public createPrimitiveTypeDouble = () : PrimitiveTypeDouble => {
		var thePrimitiveTypeDouble = new PrimitiveTypeDoubleImpl();
		
		
		PrimitiveTypeDoubleImpl.allInstances_.add(thePrimitiveTypeDouble);
		
		return thePrimitiveTypeDouble;
	}
	public createPrimitiveTypeShort = () : PrimitiveTypeShort => {
		var thePrimitiveTypeShort = new PrimitiveTypeShortImpl();
		
		
		PrimitiveTypeShortImpl.allInstances_.add(thePrimitiveTypeShort);
		
		return thePrimitiveTypeShort;
	}
	public createPrimitiveTypeFloat = () : PrimitiveTypeFloat => {
		var thePrimitiveTypeFloat = new PrimitiveTypeFloatImpl();
		
		
		PrimitiveTypeFloatImpl.allInstances_.add(thePrimitiveTypeFloat);
		
		return thePrimitiveTypeFloat;
	}
	public createPrimitiveTypeInt = () : PrimitiveTypeInt => {
		var thePrimitiveTypeInt = new PrimitiveTypeIntImpl();
		
		
		PrimitiveTypeIntImpl.allInstances_.add(thePrimitiveTypeInt);
		
		return thePrimitiveTypeInt;
	}
	public createPrimitiveTypeLong = () : PrimitiveTypeLong => {
		var thePrimitiveTypeLong = new PrimitiveTypeLongImpl();
		
		
		PrimitiveTypeLongImpl.allInstances_.add(thePrimitiveTypeLong);
		
		return thePrimitiveTypeLong;
	}
	public createPrimitiveTypeVoid = () : PrimitiveTypeVoid => {
		var thePrimitiveTypeVoid = new PrimitiveTypeVoidImpl();
		
		
		PrimitiveTypeVoidImpl.allInstances_.add(thePrimitiveTypeVoid);
		
		return thePrimitiveTypeVoid;
	}
	public createReturnStatement = () : ReturnStatement => {
		var theReturnStatement = new ReturnStatementImpl();
		
		
		ReturnStatementImpl.allInstances_.add(theReturnStatement);
		
		return theReturnStatement;
	}
	public createSingleVariableAccess = () : SingleVariableAccess => {
		var theSingleVariableAccess = new SingleVariableAccessImpl();
		
		
		SingleVariableAccessImpl.allInstances_.add(theSingleVariableAccess);
		
		return theSingleVariableAccess;
	}
	public createSingleVariableDeclaration = () : SingleVariableDeclaration => {
		var theSingleVariableDeclaration = new SingleVariableDeclarationImpl();
		
		
		SingleVariableDeclarationImpl.allInstances_.add(theSingleVariableDeclaration);
		
		return theSingleVariableDeclaration;
	}
	public createStatement = () : Statement => {
		var theStatement = new StatementImpl();
		
		
		StatementImpl.allInstances_.add(theStatement);
		
		return theStatement;
	}
	public createStringLiteral = () : StringLiteral => {
		var theStringLiteral = new StringLiteralImpl();
		
		
		StringLiteralImpl.allInstances_.add(theStringLiteral);
		
		return theStringLiteral;
	}
	public createSuperConstructorInvocation = () : SuperConstructorInvocation => {
		var theSuperConstructorInvocation = new SuperConstructorInvocationImpl();
		
		
		SuperConstructorInvocationImpl.allInstances_.add(theSuperConstructorInvocation);
		
		return theSuperConstructorInvocation;
	}
	public createSuperFieldAccess = () : SuperFieldAccess => {
		var theSuperFieldAccess = new SuperFieldAccessImpl();
		
		
		SuperFieldAccessImpl.allInstances_.add(theSuperFieldAccess);
		
		return theSuperFieldAccess;
	}
	public createSuperMethodInvocation = () : SuperMethodInvocation => {
		var theSuperMethodInvocation = new SuperMethodInvocationImpl();
		
		
		SuperMethodInvocationImpl.allInstances_.add(theSuperMethodInvocation);
		
		return theSuperMethodInvocation;
	}
	public createSwitchCase = () : SwitchCase => {
		var theSwitchCase = new SwitchCaseImpl();
		
		
		SwitchCaseImpl.allInstances_.add(theSwitchCase);
		
		return theSwitchCase;
	}
	public createSwitchStatement = () : SwitchStatement => {
		var theSwitchStatement = new SwitchStatementImpl();
		
		
		SwitchStatementImpl.allInstances_.add(theSwitchStatement);
		
		return theSwitchStatement;
	}
	public createSynchronizedStatement = () : SynchronizedStatement => {
		var theSynchronizedStatement = new SynchronizedStatementImpl();
		
		
		SynchronizedStatementImpl.allInstances_.add(theSynchronizedStatement);
		
		return theSynchronizedStatement;
	}
	public createTagElement = () : TagElement => {
		var theTagElement = new TagElementImpl();
		
		
		TagElementImpl.allInstances_.add(theTagElement);
		
		return theTagElement;
	}
	public createTextElement = () : TextElement => {
		var theTextElement = new TextElementImpl();
		
		
		TextElementImpl.allInstances_.add(theTextElement);
		
		return theTextElement;
	}
	public createThisExpression = () : ThisExpression => {
		var theThisExpression = new ThisExpressionImpl();
		
		
		ThisExpressionImpl.allInstances_.add(theThisExpression);
		
		return theThisExpression;
	}
	public createThrowStatement = () : ThrowStatement => {
		var theThrowStatement = new ThrowStatementImpl();
		
		
		ThrowStatementImpl.allInstances_.add(theThrowStatement);
		
		return theThrowStatement;
	}
	public createTryStatement = () : TryStatement => {
		var theTryStatement = new TryStatementImpl();
		
		
		TryStatementImpl.allInstances_.add(theTryStatement);
		
		return theTryStatement;
	}
	public createType = () : Type => {
		var theType = new TypeImpl();
		
		
		TypeImpl.allInstances_.add(theType);
		
		return theType;
	}
	public createTypeAccess = () : TypeAccess => {
		var theTypeAccess = new TypeAccessImpl();
		
		
		TypeAccessImpl.allInstances_.add(theTypeAccess);
		
		return theTypeAccess;
	}
	public createTypeDeclaration = () : TypeDeclaration => {
		var theTypeDeclaration = new TypeDeclarationImpl();
		
		
		TypeDeclarationImpl.allInstances_.add(theTypeDeclaration);
		
		return theTypeDeclaration;
	}
	public createTypeDeclarationStatement = () : TypeDeclarationStatement => {
		var theTypeDeclarationStatement = new TypeDeclarationStatementImpl();
		
		
		TypeDeclarationStatementImpl.allInstances_.add(theTypeDeclarationStatement);
		
		return theTypeDeclarationStatement;
	}
	public createTypeLiteral = () : TypeLiteral => {
		var theTypeLiteral = new TypeLiteralImpl();
		
		
		TypeLiteralImpl.allInstances_.add(theTypeLiteral);
		
		return theTypeLiteral;
	}
	public createTypeParameter = () : TypeParameter => {
		var theTypeParameter = new TypeParameterImpl();
		
		
		TypeParameterImpl.allInstances_.add(theTypeParameter);
		
		return theTypeParameter;
	}
	public createUnresolvedItem = () : UnresolvedItem => {
		var theUnresolvedItem = new UnresolvedItemImpl();
		
		
		UnresolvedItemImpl.allInstances_.add(theUnresolvedItem);
		
		return theUnresolvedItem;
	}
	public createUnresolvedItemAccess = () : UnresolvedItemAccess => {
		var theUnresolvedItemAccess = new UnresolvedItemAccessImpl();
		
		
		UnresolvedItemAccessImpl.allInstances_.add(theUnresolvedItemAccess);
		
		return theUnresolvedItemAccess;
	}
	public createUnresolvedAnnotationDeclaration = () : UnresolvedAnnotationDeclaration => {
		var theUnresolvedAnnotationDeclaration = new UnresolvedAnnotationDeclarationImpl();
		
		
		UnresolvedAnnotationDeclarationImpl.allInstances_.add(theUnresolvedAnnotationDeclaration);
		
		return theUnresolvedAnnotationDeclaration;
	}
	public createUnresolvedAnnotationTypeMemberDeclaration = () : UnresolvedAnnotationTypeMemberDeclaration => {
		var theUnresolvedAnnotationTypeMemberDeclaration = new UnresolvedAnnotationTypeMemberDeclarationImpl();
		
		
		UnresolvedAnnotationTypeMemberDeclarationImpl.allInstances_.add(theUnresolvedAnnotationTypeMemberDeclaration);
		
		return theUnresolvedAnnotationTypeMemberDeclaration;
	}
	public createUnresolvedClassDeclaration = () : UnresolvedClassDeclaration => {
		var theUnresolvedClassDeclaration = new UnresolvedClassDeclarationImpl();
		
		
		UnresolvedClassDeclarationImpl.allInstances_.add(theUnresolvedClassDeclaration);
		
		return theUnresolvedClassDeclaration;
	}
	public createUnresolvedEnumDeclaration = () : UnresolvedEnumDeclaration => {
		var theUnresolvedEnumDeclaration = new UnresolvedEnumDeclarationImpl();
		
		
		UnresolvedEnumDeclarationImpl.allInstances_.add(theUnresolvedEnumDeclaration);
		
		return theUnresolvedEnumDeclaration;
	}
	public createUnresolvedInterfaceDeclaration = () : UnresolvedInterfaceDeclaration => {
		var theUnresolvedInterfaceDeclaration = new UnresolvedInterfaceDeclarationImpl();
		
		
		UnresolvedInterfaceDeclarationImpl.allInstances_.add(theUnresolvedInterfaceDeclaration);
		
		return theUnresolvedInterfaceDeclaration;
	}
	public createUnresolvedLabeledStatement = () : UnresolvedLabeledStatement => {
		var theUnresolvedLabeledStatement = new UnresolvedLabeledStatementImpl();
		
		
		UnresolvedLabeledStatementImpl.allInstances_.add(theUnresolvedLabeledStatement);
		
		return theUnresolvedLabeledStatement;
	}
	public createUnresolvedMethodDeclaration = () : UnresolvedMethodDeclaration => {
		var theUnresolvedMethodDeclaration = new UnresolvedMethodDeclarationImpl();
		
		
		UnresolvedMethodDeclarationImpl.allInstances_.add(theUnresolvedMethodDeclaration);
		
		return theUnresolvedMethodDeclaration;
	}
	public createUnresolvedSingleVariableDeclaration = () : UnresolvedSingleVariableDeclaration => {
		var theUnresolvedSingleVariableDeclaration = new UnresolvedSingleVariableDeclarationImpl();
		
		
		UnresolvedSingleVariableDeclarationImpl.allInstances_.add(theUnresolvedSingleVariableDeclaration);
		
		return theUnresolvedSingleVariableDeclaration;
	}
	public createUnresolvedType = () : UnresolvedType => {
		var theUnresolvedType = new UnresolvedTypeImpl();
		
		
		UnresolvedTypeImpl.allInstances_.add(theUnresolvedType);
		
		return theUnresolvedType;
	}
	public createUnresolvedTypeDeclaration = () : UnresolvedTypeDeclaration => {
		var theUnresolvedTypeDeclaration = new UnresolvedTypeDeclarationImpl();
		
		
		UnresolvedTypeDeclarationImpl.allInstances_.add(theUnresolvedTypeDeclaration);
		
		return theUnresolvedTypeDeclaration;
	}
	public createUnresolvedVariableDeclarationFragment = () : UnresolvedVariableDeclarationFragment => {
		var theUnresolvedVariableDeclarationFragment = new UnresolvedVariableDeclarationFragmentImpl();
		
		
		UnresolvedVariableDeclarationFragmentImpl.allInstances_.add(theUnresolvedVariableDeclarationFragment);
		
		return theUnresolvedVariableDeclarationFragment;
	}
	public createVariableDeclaration = () : VariableDeclaration => {
		var theVariableDeclaration = new VariableDeclarationImpl();
		
		
		VariableDeclarationImpl.allInstances_.add(theVariableDeclaration);
		
		return theVariableDeclaration;
	}
	public createVariableDeclarationExpression = () : VariableDeclarationExpression => {
		var theVariableDeclarationExpression = new VariableDeclarationExpressionImpl();
		
		
		VariableDeclarationExpressionImpl.allInstances_.add(theVariableDeclarationExpression);
		
		return theVariableDeclarationExpression;
	}
	public createVariableDeclarationFragment = () : VariableDeclarationFragment => {
		var theVariableDeclarationFragment = new VariableDeclarationFragmentImpl();
		
		
		VariableDeclarationFragmentImpl.allInstances_.add(theVariableDeclarationFragment);
		
		return theVariableDeclarationFragment;
	}
	public createVariableDeclarationStatement = () : VariableDeclarationStatement => {
		var theVariableDeclarationStatement = new VariableDeclarationStatementImpl();
		
		
		VariableDeclarationStatementImpl.allInstances_.add(theVariableDeclarationStatement);
		
		return theVariableDeclarationStatement;
	}
	public createWildCardType = () : WildCardType => {
		var theWildCardType = new WildCardTypeImpl();
		
		
		WildCardTypeImpl.allInstances_.add(theWildCardType);
		
		return theWildCardType;
	}
	public createWhileStatement = () : WhileStatement => {
		var theWhileStatement = new WhileStatementImpl();
		
		
		WhileStatementImpl.allInstances_.add(theWhileStatement);
		
		return theWhileStatement;
	}
	
	public create(eClass:EClass):EObject {
		switch (eClass.getClassifierID()) {
			case Java_PackageImpl.ANNOTATION: return this.createAnnotation();
			case Java_PackageImpl.ARCHIVE: return this.createArchive();
			case Java_PackageImpl.ASSERTSTATEMENT: return this.createAssertStatement();
			case Java_PackageImpl.ANNOTATIONMEMBERVALUEPAIR: return this.createAnnotationMemberValuePair();
			case Java_PackageImpl.ANNOTATIONTYPEDECLARATION: return this.createAnnotationTypeDeclaration();
			case Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION: return this.createAnnotationTypeMemberDeclaration();
			case Java_PackageImpl.ANONYMOUSCLASSDECLARATION: return this.createAnonymousClassDeclaration();
			case Java_PackageImpl.ARRAYACCESS: return this.createArrayAccess();
			case Java_PackageImpl.ARRAYCREATION: return this.createArrayCreation();
			case Java_PackageImpl.ARRAYINITIALIZER: return this.createArrayInitializer();
			case Java_PackageImpl.ARRAYLENGTHACCESS: return this.createArrayLengthAccess();
			case Java_PackageImpl.ARRAYTYPE: return this.createArrayType();
			case Java_PackageImpl.ASSIGNMENT: return this.createAssignment();
			case Java_PackageImpl.BOOLEANLITERAL: return this.createBooleanLiteral();
			case Java_PackageImpl.BLOCKCOMMENT: return this.createBlockComment();
			case Java_PackageImpl.BLOCK: return this.createBlock();
			case Java_PackageImpl.BREAKSTATEMENT: return this.createBreakStatement();
			case Java_PackageImpl.CASTEXPRESSION: return this.createCastExpression();
			case Java_PackageImpl.CATCHCLAUSE: return this.createCatchClause();
			case Java_PackageImpl.CHARACTERLITERAL: return this.createCharacterLiteral();
			case Java_PackageImpl.CLASSFILE: return this.createClassFile();
			case Java_PackageImpl.CLASSINSTANCECREATION: return this.createClassInstanceCreation();
			case Java_PackageImpl.CONSTRUCTORDECLARATION: return this.createConstructorDeclaration();
			case Java_PackageImpl.CONDITIONALEXPRESSION: return this.createConditionalExpression();
			case Java_PackageImpl.CONSTRUCTORINVOCATION: return this.createConstructorInvocation();
			case Java_PackageImpl.CLASSDECLARATION: return this.createClassDeclaration();
			case Java_PackageImpl.COMPILATIONUNIT: return this.createCompilationUnit();
			case Java_PackageImpl.CONTINUESTATEMENT: return this.createContinueStatement();
			case Java_PackageImpl.DOSTATEMENT: return this.createDoStatement();
			case Java_PackageImpl.EMPTYSTATEMENT: return this.createEmptyStatement();
			case Java_PackageImpl.ENHANCEDFORSTATEMENT: return this.createEnhancedForStatement();
			case Java_PackageImpl.ENUMCONSTANTDECLARATION: return this.createEnumConstantDeclaration();
			case Java_PackageImpl.ENUMDECLARATION: return this.createEnumDeclaration();
			case Java_PackageImpl.EXPRESSIONSTATEMENT: return this.createExpressionStatement();
			case Java_PackageImpl.FIELDACCESS: return this.createFieldAccess();
			case Java_PackageImpl.FIELDDECLARATION: return this.createFieldDeclaration();
			case Java_PackageImpl.FORSTATEMENT: return this.createForStatement();
			case Java_PackageImpl.IFSTATEMENT: return this.createIfStatement();
			case Java_PackageImpl.IMPORTDECLARATION: return this.createImportDeclaration();
			case Java_PackageImpl.INFIXEXPRESSION: return this.createInfixExpression();
			case Java_PackageImpl.INITIALIZER: return this.createInitializer();
			case Java_PackageImpl.INSTANCEOFEXPRESSION: return this.createInstanceofExpression();
			case Java_PackageImpl.INTERFACEDECLARATION: return this.createInterfaceDeclaration();
			case Java_PackageImpl.JAVADOC: return this.createJavadoc();
			case Java_PackageImpl.LABELEDSTATEMENT: return this.createLabeledStatement();
			case Java_PackageImpl.LINECOMMENT: return this.createLineComment();
			case Java_PackageImpl.MANIFEST: return this.createManifest();
			case Java_PackageImpl.MANIFESTATTRIBUTE: return this.createManifestAttribute();
			case Java_PackageImpl.MANIFESTENTRY: return this.createManifestEntry();
			case Java_PackageImpl.MEMBERREF: return this.createMemberRef();
			case Java_PackageImpl.METHODDECLARATION: return this.createMethodDeclaration();
			case Java_PackageImpl.METHODINVOCATION: return this.createMethodInvocation();
			case Java_PackageImpl.METHODREF: return this.createMethodRef();
			case Java_PackageImpl.METHODREFPARAMETER: return this.createMethodRefParameter();
			case Java_PackageImpl.MODEL: return this.createModel();
			case Java_PackageImpl.MODIFIER: return this.createModifier();
			case Java_PackageImpl.NUMBERLITERAL: return this.createNumberLiteral();
			case Java_PackageImpl.NULLLITERAL: return this.createNullLiteral();
			case Java_PackageImpl.PACKAGE: return this.createPackage();
			case Java_PackageImpl.PACKAGEACCESS: return this.createPackageAccess();
			case Java_PackageImpl.PARAMETERIZEDTYPE: return this.createParameterizedType();
			case Java_PackageImpl.PARENTHESIZEDEXPRESSION: return this.createParenthesizedExpression();
			case Java_PackageImpl.POSTFIXEXPRESSION: return this.createPostfixExpression();
			case Java_PackageImpl.PREFIXEXPRESSION: return this.createPrefixExpression();
			case Java_PackageImpl.PRIMITIVETYPE: return this.createPrimitiveType();
			case Java_PackageImpl.PRIMITIVETYPEBOOLEAN: return this.createPrimitiveTypeBoolean();
			case Java_PackageImpl.PRIMITIVETYPEBYTE: return this.createPrimitiveTypeByte();
			case Java_PackageImpl.PRIMITIVETYPECHAR: return this.createPrimitiveTypeChar();
			case Java_PackageImpl.PRIMITIVETYPEDOUBLE: return this.createPrimitiveTypeDouble();
			case Java_PackageImpl.PRIMITIVETYPESHORT: return this.createPrimitiveTypeShort();
			case Java_PackageImpl.PRIMITIVETYPEFLOAT: return this.createPrimitiveTypeFloat();
			case Java_PackageImpl.PRIMITIVETYPEINT: return this.createPrimitiveTypeInt();
			case Java_PackageImpl.PRIMITIVETYPELONG: return this.createPrimitiveTypeLong();
			case Java_PackageImpl.PRIMITIVETYPEVOID: return this.createPrimitiveTypeVoid();
			case Java_PackageImpl.RETURNSTATEMENT: return this.createReturnStatement();
			case Java_PackageImpl.SINGLEVARIABLEACCESS: return this.createSingleVariableAccess();
			case Java_PackageImpl.SINGLEVARIABLEDECLARATION: return this.createSingleVariableDeclaration();
			case Java_PackageImpl.STRINGLITERAL: return this.createStringLiteral();
			case Java_PackageImpl.SUPERCONSTRUCTORINVOCATION: return this.createSuperConstructorInvocation();
			case Java_PackageImpl.SUPERFIELDACCESS: return this.createSuperFieldAccess();
			case Java_PackageImpl.SUPERMETHODINVOCATION: return this.createSuperMethodInvocation();
			case Java_PackageImpl.SWITCHCASE: return this.createSwitchCase();
			case Java_PackageImpl.SWITCHSTATEMENT: return this.createSwitchStatement();
			case Java_PackageImpl.SYNCHRONIZEDSTATEMENT: return this.createSynchronizedStatement();
			case Java_PackageImpl.TAGELEMENT: return this.createTagElement();
			case Java_PackageImpl.TEXTELEMENT: return this.createTextElement();
			case Java_PackageImpl.THISEXPRESSION: return this.createThisExpression();
			case Java_PackageImpl.THROWSTATEMENT: return this.createThrowStatement();
			case Java_PackageImpl.TRYSTATEMENT: return this.createTryStatement();
			case Java_PackageImpl.TYPEACCESS: return this.createTypeAccess();
			case Java_PackageImpl.TYPEDECLARATIONSTATEMENT: return this.createTypeDeclarationStatement();
			case Java_PackageImpl.TYPELITERAL: return this.createTypeLiteral();
			case Java_PackageImpl.TYPEPARAMETER: return this.createTypeParameter();
			case Java_PackageImpl.UNRESOLVEDITEM: return this.createUnresolvedItem();
			case Java_PackageImpl.UNRESOLVEDITEMACCESS: return this.createUnresolvedItemAccess();
			case Java_PackageImpl.UNRESOLVEDANNOTATIONDECLARATION: return this.createUnresolvedAnnotationDeclaration();
			case Java_PackageImpl.UNRESOLVEDANNOTATIONTYPEMEMBERDECLARATION: return this.createUnresolvedAnnotationTypeMemberDeclaration();
			case Java_PackageImpl.UNRESOLVEDCLASSDECLARATION: return this.createUnresolvedClassDeclaration();
			case Java_PackageImpl.UNRESOLVEDENUMDECLARATION: return this.createUnresolvedEnumDeclaration();
			case Java_PackageImpl.UNRESOLVEDINTERFACEDECLARATION: return this.createUnresolvedInterfaceDeclaration();
			case Java_PackageImpl.UNRESOLVEDLABELEDSTATEMENT: return this.createUnresolvedLabeledStatement();
			case Java_PackageImpl.UNRESOLVEDMETHODDECLARATION: return this.createUnresolvedMethodDeclaration();
			case Java_PackageImpl.UNRESOLVEDSINGLEVARIABLEDECLARATION: return this.createUnresolvedSingleVariableDeclaration();
			case Java_PackageImpl.UNRESOLVEDTYPE: return this.createUnresolvedType();
			case Java_PackageImpl.UNRESOLVEDTYPEDECLARATION: return this.createUnresolvedTypeDeclaration();
			case Java_PackageImpl.UNRESOLVEDVARIABLEDECLARATIONFRAGMENT: return this.createUnresolvedVariableDeclarationFragment();
			case Java_PackageImpl.VARIABLEDECLARATIONEXPRESSION: return this.createVariableDeclarationExpression();
			case Java_PackageImpl.VARIABLEDECLARATIONFRAGMENT: return this.createVariableDeclarationFragment();
			case Java_PackageImpl.VARIABLEDECLARATIONSTATEMENT: return this.createVariableDeclarationStatement();
			case Java_PackageImpl.WILDCARDTYPE: return this.createWildCardType();
			case Java_PackageImpl.WHILESTATEMENT: return this.createWhileStatement();
			default:
				throw new Error("The class '" + eClass.name + "' is not a valid classifier");
		}
	}
	
	
	public createFromString(eDataType:EDataType, initialValue:string):any {
		switch (eDataType.getClassifierID()) {
		case Java_PackageImpl.ASSIGNMENTKIND:
			return this.createAssignmentKindFromString(eDataType, initialValue);
		case Java_PackageImpl.INFIXEXPRESSIONKIND:
			return this.createInfixExpressionKindFromString(eDataType, initialValue);
		case Java_PackageImpl.INHERITANCEKIND:
			return this.createInheritanceKindFromString(eDataType, initialValue);
		case Java_PackageImpl.POSTFIXEXPRESSIONKIND:
			return this.createPostfixExpressionKindFromString(eDataType, initialValue);
		case Java_PackageImpl.PREFIXEXPRESSIONKIND:
			return this.createPrefixExpressionKindFromString(eDataType, initialValue);
		case Java_PackageImpl.VISIBILITYKIND:
			return this.createVisibilityKindFromString(eDataType, initialValue);
		default:
			throw new Error("The datatype '" + eDataType.name + "' is not a valid classifier");
		}
	}
	
	public convertToString(eDataType:EDataType, instanceValue:any):string {
		switch (eDataType.getClassifierID()) {
		case Java_PackageImpl.ASSIGNMENTKIND:
			return this.convertAssignmentKindToString(eDataType, instanceValue);
		case Java_PackageImpl.INFIXEXPRESSIONKIND:
			return this.convertInfixExpressionKindToString(eDataType, instanceValue);
		case Java_PackageImpl.INHERITANCEKIND:
			return this.convertInheritanceKindToString(eDataType, instanceValue);
		case Java_PackageImpl.POSTFIXEXPRESSIONKIND:
			return this.convertPostfixExpressionKindToString(eDataType, instanceValue);
		case Java_PackageImpl.PREFIXEXPRESSIONKIND:
			return this.convertPrefixExpressionKindToString(eDataType, instanceValue);
		case Java_PackageImpl.VISIBILITYKIND:
			return this.convertVisibilityKindToString(eDataType, instanceValue);
		default:
			throw new Error("The datatype '" + eDataType.name + "' is not a valid classifier");
		}
	}
	
	
	public createAssignmentKindFromString(eDataType:EDataType, initialValue:string):AssignmentKind {
		let result:AssignmentKind = AssignmentKind.get_string(initialValue);
		if (result == null)
			throw new Error(
	                        "The value '" + initialValue + "' is not a valid enumerator of '" + eDataType.name + "'");
		return result;
	}
	
	public convertAssignmentKindToString(eDataType:EDataType, instanceValue:any):string {
		return instanceValue === null ? null : instanceValue.toString();
	}
	public createInfixExpressionKindFromString(eDataType:EDataType, initialValue:string):InfixExpressionKind {
		let result:InfixExpressionKind = InfixExpressionKind.get_string(initialValue);
		if (result == null)
			throw new Error(
	                        "The value '" + initialValue + "' is not a valid enumerator of '" + eDataType.name + "'");
		return result;
	}
	
	public convertInfixExpressionKindToString(eDataType:EDataType, instanceValue:any):string {
		return instanceValue === null ? null : instanceValue.toString();
	}
	public createInheritanceKindFromString(eDataType:EDataType, initialValue:string):InheritanceKind {
		let result:InheritanceKind = InheritanceKind.get_string(initialValue);
		if (result == null)
			throw new Error(
	                        "The value '" + initialValue + "' is not a valid enumerator of '" + eDataType.name + "'");
		return result;
	}
	
	public convertInheritanceKindToString(eDataType:EDataType, instanceValue:any):string {
		return instanceValue === null ? null : instanceValue.toString();
	}
	public createPostfixExpressionKindFromString(eDataType:EDataType, initialValue:string):PostfixExpressionKind {
		let result:PostfixExpressionKind = PostfixExpressionKind.get_string(initialValue);
		if (result == null)
			throw new Error(
	                        "The value '" + initialValue + "' is not a valid enumerator of '" + eDataType.name + "'");
		return result;
	}
	
	public convertPostfixExpressionKindToString(eDataType:EDataType, instanceValue:any):string {
		return instanceValue === null ? null : instanceValue.toString();
	}
	public createPrefixExpressionKindFromString(eDataType:EDataType, initialValue:string):PrefixExpressionKind {
		let result:PrefixExpressionKind = PrefixExpressionKind.get_string(initialValue);
		if (result == null)
			throw new Error(
	                        "The value '" + initialValue + "' is not a valid enumerator of '" + eDataType.name + "'");
		return result;
	}
	
	public convertPrefixExpressionKindToString(eDataType:EDataType, instanceValue:any):string {
		return instanceValue === null ? null : instanceValue.toString();
	}
	public createVisibilityKindFromString(eDataType:EDataType, initialValue:string):VisibilityKind {
		let result:VisibilityKind = VisibilityKind.get_string(initialValue);
		if (result == null)
			throw new Error(
	                        "The value '" + initialValue + "' is not a valid enumerator of '" + eDataType.name + "'");
		return result;
	}
	
	public convertVisibilityKindToString(eDataType:EDataType, instanceValue:any):string {
		return instanceValue === null ? null : instanceValue.toString();
	}
}
