// java_/SuperMethodInvocationBase.ts
import {Java_PackageImpl} from "java_/Java_PackageImpl";
import {Set} from "ecore/Set";
import {AbstractCollection} from "ecore/AbstractCollection";
import {ENotificationImpl} from "ecore/ENotificationImpl";
import {BasicEObjectImpl} from "ecore/BasicEObjectImpl";
import {Expression} from "java_/Expression";
import {EClass} from "ecore/EClass";
import {AbstractTypeQualifiedExpressionImpl} from "java_/AbstractTypeQualifiedExpressionImpl";
import {AbstractMethodInvocationImpl} from "java_/AbstractMethodInvocationImpl";
import {InternalEObject} from "ecore/InternalEObject";
import {Java_PackageLiterals} from "java_/Java_PackageLiterals";
import {OrderedSet} from "ecore/OrderedSet";
import {EObject} from "ecore/EObject";
import {NotificationImpl} from "ecore/NotificationImpl";
import {AbstractMethodDeclaration} from "java_/AbstractMethodDeclaration";
import {TypeAccess} from "java_/TypeAccess";
import {NotificationChain} from "ecore/NotificationChain";
import {SuperMethodInvocation} from "java_/SuperMethodInvocation";
		
			export class SuperMethodInvocationBase
			extends AbstractTypeQualifiedExpressionImpl
			implements SuperMethodInvocation
			{
				private _method:AbstractMethodDeclaration = null;
				get method():AbstractMethodDeclaration{
				
					return this._method;
				}
				set method(value:AbstractMethodDeclaration) {
					if (value != this._method) {
						let msgs:NotificationChain = null;
						if (this._method != null){
							msgs = (this._method).eInverseRemove(this, Java_PackageLiterals.ABSTRACTMETHODDECLARATION_USAGES, /*AbstractMethodInvocation*/ null , msgs);
						}
						if (value != null){
							msgs = value.eInverseAdd(this, Java_PackageLiterals.ABSTRACTMETHODDECLARATION_USAGES, /*AbstractMethodInvocation*/ null, msgs);
						}
						msgs = this.basicSetMethod(value, msgs);
						if (msgs != null) {
							msgs.dispatch();
						}
					}
					else if (this.eNotificationRequired()){
						this.eNotify(new ENotificationImpl(this, NotificationImpl.SET,Java_PackageLiterals.ABSTRACTMETHODINVOCATION_METHOD , value, value));
					}
				}
				private _arguments:OrderedSet<Expression> = null;
				get arguments():OrderedSet<Expression>{
					if(this._arguments===null){
						this._arguments = new OrderedSet<Expression>(this, Java_PackageLiterals.ABSTRACTMETHODINVOCATION_ARGUMENTS, BasicEObjectImpl.EOPPOSITE_FEATURE_BASE - Java_PackageLiterals.ABSTRACTMETHODINVOCATION_ARGUMENTS);
							
					}
					return this._arguments;
					
				}
				
				
				private _typeArguments:OrderedSet<TypeAccess> = null;
				get typeArguments():OrderedSet<TypeAccess>{
					if(this._typeArguments===null){
						this._typeArguments = new OrderedSet<TypeAccess>(this, Java_PackageLiterals.ABSTRACTMETHODINVOCATION_TYPEARGUMENTS, BasicEObjectImpl.EOPPOSITE_FEATURE_BASE - Java_PackageLiterals.ABSTRACTMETHODINVOCATION_TYPEARGUMENTS);
							
					}
					return this._typeArguments;
					
				}
				
				

			
				/* do not override from BasicEObject!
				protected eStaticClass():EClass{
					//return Java_PackageLiterals.SUPERMETHODINVOCATION;
					return null;
				}
				*/
			
				public eInverseAdd(otherEnd:InternalEObject, featureID:number, msgs:NotificationChain): NotificationChain{
					switch (featureID) {
						case Java_PackageLiterals.SUPERMETHODINVOCATION_METHOD:
							if (this._method != null){
								msgs = this._method.eInverseRemove(this, Java_PackageLiterals.ABSTRACTMETHODINVOCATION_METHOD, /*AbstractMethodDeclaration*/ null, msgs);
							}
							return this.basicSetMethod(otherEnd as AbstractMethodDeclaration, msgs);
					}
					//return this.eInverseAddFromAbstractTypeQualifiedExpression(otherEnd, featureID, msgs);
					return super.eInverseAdd(otherEnd, featureID, msgs);
				}
				//public eInverseAddFromSuperMethodInvocation = this.eInverseAdd;
				
				public eInverseRemove(otherEnd:InternalEObject, featureID:number, msgs:NotificationChain):NotificationChain{
					switch (featureID) {
						case Java_PackageLiterals.SUPERMETHODINVOCATION_METHOD:
							return this.basicSetMethod(null, msgs);
					}
					//return this.eInverseRemoveFromAbstractTypeQualifiedExpression(otherEnd, featureID, msgs);
					return super.eInverseRemove(otherEnd, featureID, msgs);
				}
				
				//public eInverseRemoveFromSuperMethodInvocation = this.eInverseRemove;
			
				public basicSetMethod(newobj:AbstractMethodDeclaration, msgs:NotificationChain):NotificationChain {
					let oldobj = this._method;
					this._method = newobj;
					if (this.eNotificationRequired()) {
						let notification = new ENotificationImpl(this, NotificationImpl.SET, Java_PackageLiterals.ABSTRACTMETHODINVOCATION_METHOD, oldobj, newobj);
						if (msgs == null){
							msgs = notification;
						}
						else{
							msgs.add(notification);
						}
					}
					return msgs;
				}
				
			
				public eGet_number_boolean_boolean(featureID:number, resolve:boolean, coreType:boolean):any{
					switch (featureID) {
						case Java_PackageLiterals.SUPERMETHODINVOCATION_COMMENTS:
							return this.comments;
						case Java_PackageLiterals.SUPERMETHODINVOCATION_ORIGINALCOMPILATIONUNIT:
							return this.originalCompilationUnit;
						case Java_PackageLiterals.SUPERMETHODINVOCATION_ORIGINALCLASSFILE:
							return this.originalClassFile;
						case Java_PackageLiterals.SUPERMETHODINVOCATION_QUALIFIER:
							return this.qualifier;
						case Java_PackageLiterals.SUPERMETHODINVOCATION_METHOD:
							return this.method;
						case Java_PackageLiterals.SUPERMETHODINVOCATION_ARGUMENTS:
							return this.arguments;
						case Java_PackageLiterals.SUPERMETHODINVOCATION_TYPEARGUMENTS:
							return this.typeArguments;
					}
					//return this.eGetFromAbstractTypeQualifiedExpression(featureID, resolve, coreType);
					return super.eGet(featureID, resolve, coreType);
				}
				
				public eSet_number_any(featureID:number, newValue:any):void {
					switch (featureID) {
						case Java_PackageLiterals.SUPERMETHODINVOCATION_METHOD:
							this.method = <AbstractMethodDeclaration> newValue;
							return;
						case Java_PackageLiterals.SUPERMETHODINVOCATION_ARGUMENTS:
							this.arguments.clear();
							this.arguments.concat((newValue as AbstractCollection<EObject>).map(i => i as Expression));
							return;
						case Java_PackageLiterals.SUPERMETHODINVOCATION_TYPEARGUMENTS:
							this.typeArguments.clear();
							this.typeArguments.concat((newValue as AbstractCollection<EObject>).map(i => i as TypeAccess));
							return;
					}
					super.eSet_number_any(featureID, newValue);
				}

				public eBaseStructuralFeatureID(derivedFeatureID:number, baseClass:Function):number {
					if (baseClass instanceof AbstractMethodInvocationImpl) {
						switch (derivedFeatureID) {
							case Java_PackageImpl.SUPERMETHODINVOCATION_METHOD: return Java_PackageImpl.ABSTRACTMETHODINVOCATION_METHOD;
							case Java_PackageImpl.SUPERMETHODINVOCATION_ARGUMENTS: return Java_PackageImpl.ABSTRACTMETHODINVOCATION_ARGUMENTS;
							case Java_PackageImpl.SUPERMETHODINVOCATION_TYPEARGUMENTS: return Java_PackageImpl.ABSTRACTMETHODINVOCATION_TYPEARGUMENTS;
							default: return -1;
						}
					}
					return super.eBaseStructuralFeatureID(derivedFeatureID, baseClass);
				}
				
							
				public eDerivedStructuralFeatureID(baseFeatureID:number, baseClass:Function):number {
					if (baseClass instanceof AbstractMethodInvocationImpl) {
						switch (baseFeatureID) {
							case Java_PackageImpl.ABSTRACTMETHODINVOCATION_METHOD: return Java_PackageImpl.SUPERMETHODINVOCATION_METHOD;
							case Java_PackageImpl.ABSTRACTMETHODINVOCATION_ARGUMENTS: return Java_PackageImpl.SUPERMETHODINVOCATION_ARGUMENTS;
							case Java_PackageImpl.ABSTRACTMETHODINVOCATION_TYPEARGUMENTS: return Java_PackageImpl.SUPERMETHODINVOCATION_TYPEARGUMENTS;
							default: return -1;
						}
					}
					return super.eDerivedStructuralFeatureID(baseFeatureID, baseClass);
				}
				
			}
