// java_/TypeDeclarationStatementBase.ts
import {Set} from "ecore/Set";
import {ENotificationImpl} from "ecore/ENotificationImpl";
import {BasicEObjectImpl} from "ecore/BasicEObjectImpl";
import {EClass} from "ecore/EClass";
import {InternalEObject} from "ecore/InternalEObject";
import {Java_PackageLiterals} from "java_/Java_PackageLiterals";
import {TypeDeclarationStatement} from "java_/TypeDeclarationStatement";
import {NotificationImpl} from "ecore/NotificationImpl";
import {AbstractTypeDeclaration} from "java_/AbstractTypeDeclaration";
import {NotificationChain} from "ecore/NotificationChain";
import {StatementImpl} from "java_/StatementImpl";
		
			export class TypeDeclarationStatementBase
			extends StatementImpl
			implements TypeDeclarationStatement
			{
				private _declaration:AbstractTypeDeclaration = null;
				get declaration():AbstractTypeDeclaration{
				
					return this._declaration;
				}
				set declaration(value:AbstractTypeDeclaration) {
					if (value != this._declaration) {
						let msgs:NotificationChain = null;
						if (this._declaration != null){
							msgs = (this._declaration).eInverseRemove(this, BasicEObjectImpl.EOPPOSITE_FEATURE_BASE - Java_PackageLiterals.TYPEDECLARATIONSTATEMENT_DECLARATION, /*null*/ null , msgs);
						}
						if (value != null){
							msgs = value.eInverseAdd(this, BasicEObjectImpl.EOPPOSITE_FEATURE_BASE - Java_PackageLiterals.TYPEDECLARATIONSTATEMENT_DECLARATION, /*null*/ null, msgs);
						}
						msgs = this.basicSetDeclaration(value, msgs);
						if (msgs != null) {
							msgs.dispatch();
						}
					}
					else if (this.eNotificationRequired()){
						this.eNotify(new ENotificationImpl(this, NotificationImpl.SET,Java_PackageLiterals.TYPEDECLARATIONSTATEMENT_DECLARATION , value, value));
					}
				}

			
				/* do not override from BasicEObject!
				protected eStaticClass():EClass{
					//return Java_PackageLiterals.TYPEDECLARATIONSTATEMENT;
					return null;
				}
				*/
			
			
				public basicSetDeclaration(newobj:AbstractTypeDeclaration, msgs:NotificationChain):NotificationChain {
					let oldobj = this._declaration;
					this._declaration = newobj;
					if (this.eNotificationRequired()) {
						let notification = new ENotificationImpl(this, NotificationImpl.SET, Java_PackageLiterals.TYPEDECLARATIONSTATEMENT_DECLARATION, oldobj, newobj);
						if (msgs == null){
							msgs = notification;
						}
						else{
							msgs.add(notification);
						}
					}
					return msgs;
				}
				
			
				public eGet_number_boolean_boolean(featureID:number, resolve:boolean, coreType:boolean):any{
					switch (featureID) {
						case Java_PackageLiterals.TYPEDECLARATIONSTATEMENT_COMMENTS:
							return this.comments;
						case Java_PackageLiterals.TYPEDECLARATIONSTATEMENT_ORIGINALCOMPILATIONUNIT:
							return this.originalCompilationUnit;
						case Java_PackageLiterals.TYPEDECLARATIONSTATEMENT_ORIGINALCLASSFILE:
							return this.originalClassFile;
						case Java_PackageLiterals.TYPEDECLARATIONSTATEMENT_DECLARATION:
							return this.declaration;
					}
					//return this.eGetFromStatement(featureID, resolve, coreType);
					return super.eGet(featureID, resolve, coreType);
				}
				
				public eSet_number_any(featureID:number, newValue:any):void {
					switch (featureID) {
						case Java_PackageLiterals.TYPEDECLARATIONSTATEMENT_DECLARATION:
							this.declaration = <AbstractTypeDeclaration> newValue;
							return;
					}
					super.eSet_number_any(featureID, newValue);
				}

				
			}
