// java_/UnresolvedTypeBase.ts
import {ENotificationImpl} from "ecore/ENotificationImpl";
import {BasicEObjectImpl} from "ecore/BasicEObjectImpl";
import {EClass} from "ecore/EClass";
import {UnresolvedItemImpl} from "java_/UnresolvedItemImpl";
import {InternalEObject} from "ecore/InternalEObject";
import {Java_PackageLiterals} from "java_/Java_PackageLiterals";
import {UnresolvedType} from "java_/UnresolvedType";
import {NotificationImpl} from "ecore/NotificationImpl";
import {TypeImpl} from "java_/TypeImpl";
import {NotificationChain} from "ecore/NotificationChain";
		
			export class UnresolvedTypeBase
			extends TypeImpl
			implements UnresolvedType
			{

			
				/* do not override from BasicEObject!
				protected eStaticClass():EClass{
					//return Java_PackageLiterals.UNRESOLVEDTYPE;
					return null;
				}
				*/
			
			
			
				public eGet_number_boolean_boolean(featureID:number, resolve:boolean, coreType:boolean):any{
					switch (featureID) {
						case Java_PackageLiterals.UNRESOLVEDTYPE_COMMENTS:
							return this.comments;
						case Java_PackageLiterals.UNRESOLVEDTYPE_ORIGINALCOMPILATIONUNIT:
							return this.originalCompilationUnit;
						case Java_PackageLiterals.UNRESOLVEDTYPE_ORIGINALCLASSFILE:
							return this.originalClassFile;
						case Java_PackageLiterals.UNRESOLVEDTYPE_NAME:
							return this.name;
						case Java_PackageLiterals.UNRESOLVEDTYPE_PROXY:
							return this.proxy;
						case Java_PackageLiterals.UNRESOLVEDTYPE_USAGESINIMPORTS:
							return this.usagesInImports;
						case Java_PackageLiterals.UNRESOLVEDTYPE_USAGESINTYPEACCESS:
							return this.usagesInTypeAccess;
					}
					//return this.eGetFromType(featureID, resolve, coreType);
					return super.eGet(featureID, resolve, coreType);
				}
				
				public eSet_number_any(featureID:number, newValue:any):void {
					switch (featureID) {
					}
					super.eSet_number_any(featureID, newValue);
				}

				public eBaseStructuralFeatureID(derivedFeatureID:number, baseClass:Function):number {
					if (baseClass instanceof UnresolvedItemImpl) {
						switch (derivedFeatureID) {
							default: return -1;
						}
					}
					return super.eBaseStructuralFeatureID(derivedFeatureID, baseClass);
				}
				
							
				public eDerivedStructuralFeatureID(baseFeatureID:number, baseClass:Function):number {
					if (baseClass instanceof UnresolvedItemImpl) {
						switch (baseFeatureID) {
							default: return -1;
						}
					}
					return super.eDerivedStructuralFeatureID(baseFeatureID, baseClass);
				}
				
			}
